diff --git gcc-15.2.0-old/gcc/builtins.cc gcc-15.2.0/gcc/builtins.cc
index 05201424433..d153b22f4e4 100644
--- gcc-15.2.0-old/gcc/builtins.cc
+++ gcc-15.2.0/gcc/builtins.cc
@@ -6631,7 +6631,6 @@ expand_builtin_sync_lock_release (machine_mode mode, tree exp)
 {
   /* Miosix patch begin: Generate the atomic store the long way because
      just calling expand_atomic_store does not guarantee we generate any code */
-  rtx mem = get_builtin_sync_mem (CALL_EXPR_ARG (exp, 0), mode);
   unsigned int bytes_log2 = exact_log2 (GET_MODE_SIZE (mode).to_constant ());
   gcc_assert (bytes_log2 < 5);
   built_in_function fncode
diff --git gcc-15.2.0-old/gcc/config.gcc gcc-15.2.0/gcc/config.gcc
index ba076f14aff..d38ecf8f92a 100644
--- gcc-15.2.0-old/gcc/config.gcc
+++ gcc-15.2.0/gcc/config.gcc
@@ -1507,7 +1507,7 @@ arm*-*-eabi* | arm*-*-symbianelf* | arm*-*-rtems* | arm*-*-fuchsia*)
 	arm*-miosix-eabi*)
 	  # Copy options from arm*-*-eabi*, add the miosix-specific ones
 	  # and make sure arm/t-arm-elf is not added to tmake_file
-	  tm_file="${tm_file} newlib-stdint.h arm/miosix-eabi.h"  # Append
+	  tm_file="${tm_file} newlib-stdint.h miosix.h"           # Append
 	  tmake_file="arm/t-arm arm/t-arm-miosix arm/t-bpabi"     # Replace
 	  extra_options="${extra_options} miosix.opt"             # Append
 	  use_gcc_stdint=wrap
diff --git gcc-15.2.0-old/gcc/config/arm/miosix-eabi.h gcc-15.2.0/gcc/config/arm/miosix-eabi.h
deleted file mode 100644
index 96032166306..00000000000
--- gcc-15.2.0-old/gcc/config/arm/miosix-eabi.h
+++ gcc-15.2.0/gcc/config/arm/miosix-eabi.h
@@ -1,17 +0,0 @@
-
-/*
- * RATIONALE: adding builtin_define to always define _MIOSIX
- * - when libgcc/libstdc++/newlib are compiled, as there are some #ifdef _MIOSIX
- * - when Miosix processes are compiled, to allow #ifdef _MIOSIX
- * Also add versioning to miosix-specific compiler patches.
- */
-
-#undef TARGET_OS_CPP_BUILTINS
-#define TARGET_OS_CPP_BUILTINS()         \
-    do {                                 \
-        TARGET_BPABI_CPP_BUILTINS();     \
-        builtin_define("_MIOSIX");       \
-        builtin_define("_MIOSIX_GCC_PATCH_MAJOR=4"); \
-        builtin_define("_MIOSIX_GCC_PATCH_MINOR=0"); \
-        builtin_assert("system=miosix"); \
-    } while(false)
diff --git gcc-15.2.0-old/gcc/config/arm/t-arm-miosix gcc-15.2.0/gcc/config/arm/t-arm-miosix
index 32da3627f1a..6d2b6e5470c 100644
--- gcc-15.2.0-old/gcc/config/arm/t-arm-miosix
+++ gcc-15.2.0/gcc/config/arm/t-arm-miosix
@@ -33,38 +33,38 @@ MULTILIB_DIRNAMES   += v4t v6-m v7-m v7e-m v7e-m+fp v7e-m+dp v8-m.base v8-m.main
 MULTILIB_OPTIONS    += mfloat-abi=soft/mfloat-abi=hard
 MULTILIB_DIRNAMES   += nofp hard
 
-MULTILIB_OPTIONS    += fpie msingle-pic-base
-MULTILIB_DIRNAMES   += pie single-pic-base
+MULTILIB_OPTIONS    += qkernelspace
+MULTILIB_DIRNAMES   += kernel
 
 ## Multilibs to build:
 
 # Armv4 (ARM7TDMI, no FP)
-MULTILIB_REQUIRED	+= marm/march=armv4t/mfloat-abi=soft
-MULTILIB_REQUIRED	+= mthumb/march=armv4t/mfloat-abi=soft
+MULTILIB_REQUIRED	+= marm/march=armv4t/mfloat-abi=soft/qkernelspace
+MULTILIB_REQUIRED	+= mthumb/march=armv4t/mfloat-abi=soft/qkernelspace
 
 # Armv6 (Cortex-M0, M0+, M1, no FP)
+MULTILIB_REQUIRED	+= mthumb/march=armv6s-m/mfloat-abi=soft/qkernelspace
 MULTILIB_REQUIRED	+= mthumb/march=armv6s-m/mfloat-abi=soft
-MULTILIB_REQUIRED	+= mthumb/march=armv6s-m/mfloat-abi=soft/fpie/msingle-pic-base
 
 # Armv7-M (Cortex-M3, M4, M7, no FP)
+MULTILIB_REQUIRED	+= mthumb/march=armv7-m/mfloat-abi=soft/qkernelspace
 MULTILIB_REQUIRED	+= mthumb/march=armv7-m/mfloat-abi=soft
-MULTILIB_REQUIRED	+= mthumb/march=armv7-m/mfloat-abi=soft/fpie/msingle-pic-base
 
 # Armv7e-M (Cortex-M4, M7 with FP single and double precision)
+MULTILIB_REQUIRED	+= mthumb/march=armv7e-m+fp/mfloat-abi=hard/qkernelspace
 MULTILIB_REQUIRED	+= mthumb/march=armv7e-m+fp/mfloat-abi=hard
-MULTILIB_REQUIRED	+= mthumb/march=armv7e-m+fp/mfloat-abi=hard/fpie/msingle-pic-base
+MULTILIB_REQUIRED	+= mthumb/march=armv7e-m+fp.dp/mfloat-abi=hard/qkernelspace
 MULTILIB_REQUIRED	+= mthumb/march=armv7e-m+fp.dp/mfloat-abi=hard
-MULTILIB_REQUIRED	+= mthumb/march=armv7e-m+fp.dp/mfloat-abi=hard/fpie/msingle-pic-base
 
 # Armv8-m baseline (Cortex-M23, no FP)
+MULTILIB_REQUIRED	+= mthumb/march=armv8-m.base/mfloat-abi=soft/qkernelspace
 MULTILIB_REQUIRED	+= mthumb/march=armv8-m.base/mfloat-abi=soft
-MULTILIB_REQUIRED	+= mthumb/march=armv8-m.base/mfloat-abi=soft/fpie/msingle-pic-base
 
 # Armv8-m mainline (Cortex-M33, M35P with FP single and double precision)
+MULTILIB_REQUIRED	+= mthumb/march=armv8-m.main+fp/mfloat-abi=hard/qkernelspace
 MULTILIB_REQUIRED	+= mthumb/march=armv8-m.main+fp/mfloat-abi=hard
-MULTILIB_REQUIRED	+= mthumb/march=armv8-m.main+fp/mfloat-abi=hard/fpie/msingle-pic-base
+MULTILIB_REQUIRED	+= mthumb/march=armv8-m.main+fp.dp/mfloat-abi=hard/qkernelspace
 MULTILIB_REQUIRED	+= mthumb/march=armv8-m.main+fp.dp/mfloat-abi=hard
-MULTILIB_REQUIRED	+= mthumb/march=armv8-m.main+fp.dp/mfloat-abi=hard/fpie/msingle-pic-base
 
 ## Aliases
 
@@ -95,7 +95,6 @@ MULTILIB_MATCHES	+= march?armv7e-m+fp=march?armv7e-m+fpv5
 MULTILIB_REUSE		+= mthumb/march.armv4t/mfloat-abi.soft=mthumb/march.armv4t
 MULTILIB_REUSE		+= marm/march.armv4t/mfloat-abi.soft=marm/march.armv4t
 MULTILIB_REUSE		+= $(foreach ARCH, armv6s-m armv7-m armv8-m\.base, \
-			mthumb/march.$(ARCH)/mfloat-abi.soft=mthumb/march.$(ARCH))
+			mthumb/march.$(ARCH)/mfloat-abi.soft/qkernelspace=mthumb/march.$(ARCH)/qkernelspace)
 MULTILIB_REUSE		+= $(foreach ARCH, armv6s-m armv7-m armv8-m\.base, \
-			mthumb/march.$(ARCH)/mfloat-abi.soft/fpie/msingle-pic-base=mthumb/march.$(ARCH)/fpie/msingle-pic-base)
-
+			mthumb/march.$(ARCH)/mfloat-abi.soft=mthumb/march.$(ARCH))
diff --git gcc-15.2.0-old/gcc/config/miosix.h gcc-15.2.0/gcc/config/miosix.h
new file mode 100644
index 00000000000..d7a16ac8c87
--- gcc-15.2.0-old/gcc/config/miosix.h
+++ gcc-15.2.0/gcc/config/miosix.h
@@ -0,0 +1,56 @@
+/* Target definitions for the Miosix RTOS. */
+
+/* RATIONALE: adding builtin_define to always define _MIOSIX
+   - when libgcc/libstdc++/newlib are compiled, as there are some #ifdef _MIOSIX
+   - when Miosix processes are compiled, to allow #ifdef _MIOSIX
+   Also add versioning to miosix-specific compiler patches, and specs for
+   making GCC compile Miosix processes by default. */
+
+#undef  TARGET_OS_CPP_BUILTINS
+#define TARGET_OS_CPP_BUILTINS()                    \
+  do                                                \
+    {                                               \
+      TARGET_BPABI_CPP_BUILTINS ();                 \
+      builtin_define ("_MIOSIX");                   \
+      builtin_define ("_MIOSIX_GCC_PATCH_MAJOR=4"); \
+      builtin_define ("_MIOSIX_GCC_PATCH_MINOR=0"); \
+      builtin_assert ("system=miosix");             \
+    }                                               \
+  while (0)
+
+/* Force -fpie -msingle-pic-base for Miosix process object files */
+#undef  OS_CC1_SPEC
+#define OS_CC1_SPEC "%{!qkernelspace:-fpie -msingle-pic-base}"
+
+/* No crti.o/crtbegin.o/crt0.o */
+#undef  STARTFILE_SPEC
+#define STARTFILE_SPEC ""
+
+/* No crtend.o/crtn.o */
+#undef  ENDFILE_SPEC
+#define ENDFILE_SPEC ""
+
+/* Library specs for Miosix processes */
+
+/* STD_LIB_SPEC is the default value of LIB_SPEC as defined in gcc.cc, what it
+   does is to expand to -lc in most cases except if profiling is enabled. */
+#define STD_LIB_SPEC "%{!shared:%{g*:-lg} %{!p:%{!pg:-lc}}%{p:-lc_p}%{pg:-lc_p}}"
+
+/* Includes libsyscalls if -qkernelspace is not specified.
+   Note that this spec goes unused if -nostlib is passed to gcc (i.e. we don't
+   have to handle that case manually ourselves. */
+#undef  LIB_SPEC
+#define LIB_SPEC "%{!qkernelspace:-lsyscalls} " STD_LIB_SPEC
+
+/* Linker specs for Miosix processes */
+
+/* Always remove -lpthread because its functions are already defined either by
+   the rest of the kernel, or by libsyscalls for processes.
+   Also add required options for processes to the linker's command line if
+   -qkernelspace is not specified. Note that the default LINK_SPEC is empty. */
+#undef  LINK_SPEC
+#define LINK_SPEC \
+  "%:remove-outfile(-lpthread) \
+   %{!qkernelspace: \
+     -n -pie --spare-dynamic-tags 3 --target2=mx-data-rel \
+     -T miosix_process.ld}"
diff --git gcc-15.2.0-old/gcc/config/miosix.opt gcc-15.2.0/gcc/config/miosix.opt
index b1d5f68c324..916562eb73f 100644
--- gcc-15.2.0-old/gcc/config/miosix.opt
+++ gcc-15.2.0/gcc/config/miosix.opt
@@ -5,4 +5,7 @@
 pthread
 Ignore
 
+qkernelspace
+Driver
+
 ; This comment is to ensure we retain the blank line above.
