diff -ruN newlib-3.1.0-old/newlib/libc/posix/execl.c newlib-3.1.0/newlib/libc/posix/execl.c
--- newlib-3.1.0-old/newlib/libc/posix/execl.c	2025-11-04 09:36:58.832414940 +0100
+++ newlib-3.1.0/newlib/libc/posix/execl.c	2025-11-04 09:36:35.444415711 +0100
@@ -19,6 +19,9 @@
 
 #include <stdarg.h>
 
+// Add missing declaration when compiling for newer C standards
+int _execve (const char *filename, char *const argv[], char *const envp[]);
+
 int
 execl (const char *path,
       const char *arg0, ...)
diff -ruN newlib-3.1.0-old/newlib/libc/posix/execle.c newlib-3.1.0/newlib/libc/posix/execle.c
--- newlib-3.1.0-old/newlib/libc/posix/execle.c	2025-11-04 09:36:58.832414940 +0100
+++ newlib-3.1.0/newlib/libc/posix/execle.c	2025-11-04 09:36:35.448415711 +0100
@@ -14,6 +14,9 @@
 
 #include <stdarg.h>
 
+// Add missing declaration when compiling for newer C standards
+int _execve (const char *filename, char *const argv[], char *const envp[]);
+
 int
 execle (const char *path,
       const char *arg0, ...)
diff -ruN newlib-3.1.0-old/newlib/libc/posix/execv.c newlib-3.1.0/newlib/libc/posix/execv.c
--- newlib-3.1.0-old/newlib/libc/posix/execv.c	2019-01-01 05:40:11.000000000 +0100
+++ newlib-3.1.0/newlib/libc/posix/execv.c	2025-11-04 09:36:35.448415711 +0100
@@ -13,6 +13,9 @@
    'environ'.  */
 static char ***p_environ = &environ;
 
+// Add missing declaration when compiling for newer C standards
+int _execve (const char *filename, char *const argv[], char *const envp[]);
+
 int
 execv (const char *path,
 	char * const argv[])
diff -ruN newlib-3.1.0-old/newlib/libc/posix/execve.c newlib-3.1.0/newlib/libc/posix/execve.c
--- newlib-3.1.0-old/newlib/libc/posix/execve.c	2019-01-01 05:40:11.000000000 +0100
+++ newlib-3.1.0/newlib/libc/posix/execve.c	2025-11-04 09:36:35.448415711 +0100
@@ -8,6 +8,8 @@
 
 #include <unistd.h>
 
+// Add missing declaration when compiling for newer C standards
+int _execve (const char *filename, char *const argv[], char *const envp[]);
 
 int
 execve (const char *path,
diff -ruN newlib-3.1.0-old/newlib/libc/posix/glob.c newlib-3.1.0/newlib/libc/posix/glob.c
--- newlib-3.1.0-old/newlib/libc/posix/glob.c	2019-01-01 05:40:11.000000000 +0100
+++ newlib-3.1.0/newlib/libc/posix/glob.c	2025-11-04 09:44:35.000000000 +0100
@@ -617,7 +617,7 @@
 	 * and dirent.h as taking pointers to differently typed opaque
 	 * structures.
 	 */
-	struct dirent *(*readdirfunc)();
+	struct dirent *(*readdirfunc)(void *);
 
 	if (pathend > pathend_last)
 		return (1);
@@ -642,7 +642,7 @@
 	if (pglob->gl_flags & GLOB_ALTDIRFUNC)
 		readdirfunc = pglob->gl_readdir;
 	else
-		readdirfunc = readdir;
+		readdirfunc = (struct dirent *(*)(void *))readdir;
 	while ((dp = (*readdirfunc)(dirp))) {
 		u_char *sc;
 		Char *dc;
diff -ruN newlib-3.1.0-old/newlib/libc/posix/rewinddir.c newlib-3.1.0/newlib/libc/posix/rewinddir.c
--- newlib-3.1.0-old/newlib/libc/posix/rewinddir.c	2025-11-04 09:36:58.832414940 +0100
+++ newlib-3.1.0/newlib/libc/posix/rewinddir.c	2025-11-04 09:36:35.448415711 +0100
@@ -41,6 +41,9 @@
 #include <dirent.h>
 #include <sys/lock.h>
 
+// Needed by lseek in the _MIOSIX case
+#include <unistd.h>
+
 void
 rewinddir (DIR *dirp)
 {
diff -ruN newlib-3.1.0-old/newlib/libc/posix/seekdir.c newlib-3.1.0/newlib/libc/posix/seekdir.c
--- newlib-3.1.0-old/newlib/libc/posix/seekdir.c	2019-01-01 05:40:11.000000000 +0100
+++ newlib-3.1.0/newlib/libc/posix/seekdir.c	2025-11-04 09:36:35.448415711 +0100
@@ -38,6 +38,8 @@
 #endif /* LIBC_SCCS and not lint */
 
 #include <sys/param.h>
+// _seekdir is gated behing _COMPILING_NEWLIB which is undefined (but shouldn't)
+#define _COMPILING_NEWLIB
 #include <dirent.h>
 #include <sys/lock.h>
 
