diff -Naur Include.orig/stm32f205xx.h Include/stm32f205xx.h
--- Include.orig/stm32f205xx.h	2023-04-28 12:30:55
+++ Include/stm32f205xx.h	2024-01-09 15:34:59
@@ -157,7 +157,7 @@
   * @}
   */
 
-#include "core_cm3.h"
+#include <CMSIS/Include/core_cm3.h> //By TFT: fix path
 #include "system_stm32f2xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32f207xx.h Include/stm32f207xx.h
--- Include.orig/stm32f207xx.h	2023-04-28 12:30:55
+++ Include/stm32f207xx.h	2024-01-09 15:34:59
@@ -160,7 +160,7 @@
   * @}
   */
 
-#include "core_cm3.h"
+#include <CMSIS/Include/core_cm3.h> //By TFT: fix path
 #include "system_stm32f2xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32f215xx.h Include/stm32f215xx.h
--- Include.orig/stm32f215xx.h	2023-04-28 12:30:55
+++ Include/stm32f215xx.h	2024-01-09 15:34:59
@@ -158,7 +158,7 @@
   * @}
   */
 
-#include "core_cm3.h"
+#include <CMSIS/Include/core_cm3.h> //By TFT: fix path
 #include "system_stm32f2xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32f217xx.h Include/stm32f217xx.h
--- Include.orig/stm32f217xx.h	2023-04-28 12:30:55
+++ Include/stm32f217xx.h	2024-01-09 15:34:59
@@ -161,7 +161,7 @@
   * @}
   */
 
-#include "core_cm3.h"
+#include <CMSIS/Include/core_cm3.h> //By TFT: fix path
 #include "system_stm32f2xx.h"
 #include <stdint.h>
 
