diff -ruN gcc-14.2.0-old/libgomp/configure.tgt gcc-14.2.0/libgomp/configure.tgt
--- gcc-14.2.0-old/libgomp/configure.tgt	2025-09-09 16:52:42.186587531 +0200
+++ gcc-14.2.0/libgomp/configure.tgt	2025-09-09 16:55:11.395007822 +0200
@@ -174,6 +174,8 @@
 
   *-miosix-*)
 	config_path="posix"
+	# libgomp uses pthread_exit which on Miosix throws a terminate exception
+	XCFLAGS="${XCFLAGS} -fexceptions"
 	;;
 
   *-*-rtems*)
diff -ruN gcc-14.2.0-old/libgomp/libgomp.h gcc-14.2.0/libgomp/libgomp.h
--- gcc-14.2.0-old/libgomp/libgomp.h	2025-09-09 16:52:42.118587339 +0200
+++ gcc-14.2.0/libgomp/libgomp.h	2025-09-09 16:52:53.446619244 +0200
@@ -967,9 +967,22 @@
 }
 #else
 extern pthread_key_t gomp_tls_key;
+struct gomp_thread *gomp_initialize_tls_key (void);
 static inline struct gomp_thread *gomp_thread (void)
 {
-  return pthread_getspecific (gomp_tls_key);
+  /* NOTE: if openmp is only used by main(), then initialize_team() takes care
+   * of initializing gomp_tls_key for main() as team leader. However, if the
+   * application spawns a thread using pthread, C11 or C++11 threads and *that*
+   * thread tries to call some OpenMP-parallelized code, no one initialized
+   * gomp_tls_key for that thread, and pthread_getspecific() returns NULL.
+   * Calling gomp_initialize_tls_key() solves the issue.
+   */
+  void *gomp_tls_ptr = pthread_getspecific (gomp_tls_key);
+  if (!gomp_tls_ptr)
+  {
+    gomp_tls_ptr = gomp_initialize_tls_key ();
+  }
+  return gomp_tls_ptr;
 }
 #endif
 
diff -ruN gcc-14.2.0-old/libgomp/team.c gcc-14.2.0/libgomp/team.c
--- gcc-14.2.0-old/libgomp/team.c	2025-09-09 16:52:42.114587328 +0200
+++ gcc-14.2.0/libgomp/team.c	2025-09-09 17:04:23.564563212 +0200
@@ -43,6 +43,20 @@
 __thread struct gomp_thread gomp_tls_data;
 #else
 pthread_key_t gomp_tls_key;
+
+struct gomp_thread *gomp_initialize_tls_key (void)
+{
+	struct gomp_thread *gomp_tls_ptr = malloc (sizeof (struct gomp_thread));
+	if (!gomp_tls_ptr) gomp_fatal ("Out of memory");
+	memset (gomp_tls_ptr, 0, sizeof (struct gomp_thread));
+	pthread_setspecific (gomp_tls_key, gomp_tls_ptr);
+	return gomp_tls_ptr;
+}
+
+void gomp_delete_tls_key (void* ptr)
+{
+	free (ptr);
+}
 #endif
 
 
@@ -78,8 +92,7 @@
 #if defined HAVE_TLS || defined USE_EMUTLS
   thr = &gomp_tls_data;
 #else
-  struct gomp_thread local_thr;
-  thr = &local_thr;
+  thr = gomp_initialize_tls_key ();
 #endif
   gomp_sem_init (&thr->release, 0);
 
@@ -95,9 +108,10 @@
 #ifdef GOMP_NEEDS_THREAD_HANDLE
   thr->handle = data->handle;
 #endif
-#if !(defined HAVE_TLS || defined USE_EMUTLS)
-  pthread_setspecific (gomp_tls_key, thr);
-#endif
+//TODO: code changed upstream and this line was moved down, check the reason
+// #if !(defined HAVE_TLS || defined USE_EMUTLS)
+//   pthread_setspecific (gomp_tls_key, thr);
+// #endif
 
   thr->ts.team->ordered_release[thr->ts.team_id] = &thr->release;
 
@@ -262,9 +276,13 @@
 /* Free a thread pool and release its threads. */
 
 void
-gomp_free_thread (void *arg __attribute__((unused)))
+gomp_free_thread (void *arg)
 {
-  struct gomp_thread *thr = gomp_thread ();
+  // NOTE: we cannot call gomp_thread () here as it calls pthread_getspecific ()
+  // and the POSIX semantics require the pthread_key value to be set to NULL
+  // before calling the destructor. We must get the struct gomp_thread pointer
+  // from the arg parameter!
+  struct gomp_thread *thr = (struct gomp_thread *) arg;
   struct gomp_thread_pool *pool = thr->thread_pool;
   if (pool)
     {
@@ -1022,10 +1040,8 @@
 initialize_team (void)
 {
 #if !defined HAVE_TLS && !defined USE_EMUTLS
-  static struct gomp_thread initial_thread_tls_data;
-
-  pthread_key_create (&gomp_tls_key, NULL);
-  pthread_setspecific (gomp_tls_key, &initial_thread_tls_data);
+  pthread_key_create (&gomp_tls_key, gomp_delete_tls_key);
+  gomp_initialize_tls_key ();
 #endif
 
   if (pthread_key_create (&gomp_thread_destructor, gomp_free_thread) != 0)
