diff -ruN gcc-14.2.0-orig/libgomp/libgomp.h gcc-14.2.0/libgomp/libgomp.h
--- gcc-14.2.0-orig/libgomp/libgomp.h	2024-08-01 10:17:18.000000000 +0200
+++ gcc-14.2.0/libgomp/libgomp.h	2025-04-23 16:34:43.247701687 +0200
@@ -966,9 +966,27 @@
   return &gomp_tls_data;
 }
 #else
+extern void initialize_team (void);
 extern pthread_key_t gomp_tls_key;
 static inline struct gomp_thread *gomp_thread (void)
 {
+  /* NOTE: initialize_team() is declared __attribute__((constructor)) and sets
+   * gomp_tls_key at program start, but for main() only! However, if the
+   * application spawns a thread using pthread, C11 or C++11 threads and *that*
+   * thread tries to call some OpenMP-parallelized code, no one initialized
+   * gomp_tls_key for that thread, causing a NULL pointer dereference.
+   * Fix: call initialize_team() if we reach here and gomp_tls_key is NULL
+   */
+  void *gomp_tls_ptr = pthread_getspecific (gomp_tls_key);
+  if (!gomp_tls_ptr)
+  {
+    initialize_team ();
+    gomp_tls_ptr = pthread_getspecific (gomp_tls_key);
+  }
+  return gomp_tls_ptr;
+}
+static inline struct gomp_thread *gomp_thread_no_init (void)
+{
   return pthread_getspecific (gomp_tls_key);
 }
 #endif
diff -ruN gcc-14.2.0-orig/libgomp/team.c gcc-14.2.0/libgomp/team.c
--- gcc-14.2.0-orig/libgomp/team.c	2024-08-01 10:17:18.000000000 +0200
+++ gcc-14.2.0/libgomp/team.c	2025-04-23 16:34:43.249889195 +0200
@@ -264,7 +264,12 @@
 void
 gomp_free_thread (void *arg __attribute__((unused)))
 {
+#ifdef LIBGOMP_USE_PTHREADS
+  struct gomp_thread *thr = gomp_thread_no_init ();
+  if (!thr) return;
+#else
   struct gomp_thread *thr = gomp_thread ();
+#endif
   struct gomp_thread_pool *pool = thr->thread_pool;
   if (pool)
     {
@@ -1018,7 +1023,7 @@
 
 /* Constructors for this file.  */
 
-static void __attribute__((constructor))
+void __attribute__((constructor))
 initialize_team (void)
 {
 #if !defined HAVE_TLS && !defined USE_EMUTLS
