--- gcc-4.5.2/libstdc++-v3/libsupc++/eh_alloc.cc	2009-04-09 17:00:19.000000000 +0200
+++ gcc-4.5.2/libstdc++-v3/libsupc++/eh_alloc_patched.cc	2011-04-10 20:18:36.676682465 +0200
@@ -73,6 +73,15 @@
 # define EMERGENCY_OBJ_COUNT	4
 #endif
 
+//begin TFT patch
+//RATIONALE: reduced emergency buffer for ARM microcontrollers, saves ~1.5K of RAM
+#undef EMERGENCY_OBJ_SIZE
+#undef EMERGENCY_OBJ_COUNT
+
+#define EMERGENCY_OBJ_SIZE  160
+#define EMERGENCY_OBJ_COUNT 4
+//end TFT patch
+
 #if INT_MAX == 32767 || EMERGENCY_OBJ_COUNT <= 32
 typedef unsigned int bitmask_type;
 #else
