diff -ruN gcc-4.5.2-original/gcc/config/arm/t-arm-elf gcc-4.5.2/gcc/config/arm/t-arm-elf
--- gcc-4.5.2-original/gcc/config/arm/t-arm-elf	2009-08-06 16:27:45.000000000 +0200
+++ gcc-4.5.2/gcc/config/arm/t-arm-elf	2011-05-30 10:35:06.110193855 +0200
@@ -36,15 +36,20 @@
 MULTILIB_EXCEPTIONS  = 
 MULTILIB_MATCHES     =
 
-#MULTILIB_OPTIONS      += march=armv7
-#MULTILIB_DIRNAMES     += thumb2
-#MULTILIB_EXCEPTIONS   += march=armv7* marm/*march=armv7*
-#MULTILIB_MATCHES      += march?armv7=march?armv7-a
-#MULTILIB_MATCHES      += march?armv7=march?armv7-r
-#MULTILIB_MATCHES      += march?armv7=march?armv7-m
-#MULTILIB_MATCHES      += march?armv7=mcpu?cortex-a8
-#MULTILIB_MATCHES      += march?armv7=mcpu?cortex-r4
-#MULTILIB_MATCHES      += march?armv7=mcpu?cortex-m3
+## begin TFT patch
+## RATIONALE: building also thumb2 multilibs allow supporting Cortex-M3 processors
+## as well as ARM7 processors with the same compiler
+## see http://gcc.gnu.org/ml/gcc-help/2009-02/msg00033.html
+MULTILIB_OPTIONS      += march=armv7
+MULTILIB_DIRNAMES     += thumb2
+MULTILIB_EXCEPTIONS   += march=armv7* marm/*march=armv7*
+MULTILIB_MATCHES      += march?armv7=march?armv7-a
+MULTILIB_MATCHES      += march?armv7=march?armv7-r
+MULTILIB_MATCHES      += march?armv7=march?armv7-m
+MULTILIB_MATCHES      += march?armv7=mcpu?cortex-a8
+MULTILIB_MATCHES      += march?armv7=mcpu?cortex-r4
+MULTILIB_MATCHES      += march?armv7=mcpu?cortex-m3
+## end TFT patch
 
 # Not quite true.  We can support hard-vfp calling in Thumb2, but how do we
 # express that here?  Also, we really need architecture v5e or later
diff -ruN gcc-4.5.2-original/gcc/config.gcc gcc-4.5.2/gcc/config.gcc
--- gcc-4.5.2-original/gcc/config.gcc	2010-10-09 11:35:53.000000000 +0200
+++ gcc-4.5.2/gcc/config.gcc	2011-05-30 10:35:06.435193861 +0200
@@ -3512,6 +3512,10 @@
 		;;
 esac
 
+case ${enable_threads} in
+    miosix) thread_file='miosix' ;;
+esac
+
 t=
 all_defaults="abi cpu cpu_32 cpu_64 arch arch_32 arch_64 tune tune_32 tune_64 schedule float mode fpu divide llsc mips-plt synci"
 for option in $all_defaults
diff -ruN gcc-4.5.2-original/gcc/configure gcc-4.5.2/gcc/configure
--- gcc-4.5.2-original/gcc/configure	2010-12-03 13:35:37.000000000 +0100
+++ gcc-4.5.2/gcc/configure	2011-05-30 10:35:06.444193873 +0200
@@ -10421,7 +10421,7 @@
     target_thread_file='single'
     ;;
   aix | dce | gnat | irix | posix | posix95 | rtems | \
-  single | solaris | vxworks | win32 | mipssde)
+  single | solaris | vxworks | win32 | mipssde | miosix)
     target_thread_file=${enable_threads}
     ;;
   *)
diff -ruN gcc-4.5.2-original/gcc/configure.ac gcc-4.5.2/gcc/configure.ac
--- gcc-4.5.2-original/gcc/configure.ac	2010-12-03 13:35:37.000000000 +0100
+++ gcc-4.5.2/gcc/configure.ac	2011-05-30 10:35:06.616194078 +0200
@@ -1373,7 +1373,7 @@
     target_thread_file='single'
     ;;
   aix | dce | gnat | irix | posix | posix95 | rtems | \
-  single | solaris | vxworks | win32 | mipssde)
+  single | solaris | vxworks | win32 | mipssde | miosix)
     target_thread_file=${enable_threads}
     ;;
   *)
diff -ruN gcc-4.5.2-original/gcc/gthr.h gcc-4.5.2/gcc/gthr.h
--- gcc-4.5.2-original/gcc/gthr.h	2009-11-25 11:55:54.000000000 +0100
+++ gcc-4.5.2/gcc/gthr.h	2011-05-30 10:35:06.617194051 +0200
@@ -140,31 +140,35 @@
 
 */
 
-/* Check first for thread specific defines.  */
-#if defined (__tpf__)
-#include "gthr-tpf.h"
-#elif _PTHREADS
-#include "gthr-posix.h"
-#elif _PTHREADS95
-#include "gthr-posix95.h"
-#elif _DCE_THREADS
-#include "gthr-dce.h"
-#elif _SOLARIS_THREADS
-#include "gthr-solaris.h"
-
-/* Include GTHREAD_FILE if one is defined.  */
-#elif defined(HAVE_GTHR_DEFAULT)
-#if SUPPORTS_WEAK
-#ifndef GTHREAD_USE_WEAK
-#define GTHREAD_USE_WEAK 1
-#endif
-#endif
-#include "gthr-default.h"
-
-/* Fallback to single thread definitions.  */
-#else
-#include "gthr-single.h"
-#endif
+// /* Check first for thread specific defines.  */
+// #if defined (__tpf__)
+// #include "gthr-tpf.h"
+// #elif _PTHREADS
+// #include "gthr-posix.h"
+// #elif _PTHREADS95
+// #include "gthr-posix95.h"
+// #elif _DCE_THREADS
+// #include "gthr-dce.h"
+// #elif _SOLARIS_THREADS
+// #include "gthr-solaris.h"
+// 
+// /* Include GTHREAD_FILE if one is defined.  */
+// #elif defined(HAVE_GTHR_DEFAULT)
+// #if SUPPORTS_WEAK
+// #ifndef GTHREAD_USE_WEAK
+// #define GTHREAD_USE_WEAK 1
+// #endif
+// #endif
+// #include "gthr-default.h"
+// 
+// /* Fallback to single thread definitions.  */
+// #else
+// #include "gthr-single.h"
+// #endif
+
+//TODO: actually here there should be an #elif _MIOSIX but for now
+//it's easier to remove all those #ifdefs and force gthr-miosix.h
+#include "gthr-miosix.h"
 
 #ifndef HIDE_EXPORTS
 #pragma GCC visibility pop
diff -ruN gcc-4.5.2-original/gcc/gthr-miosix.h gcc-4.5.2/gcc/gthr-miosix.h
--- gcc-4.5.2-original/gcc/gthr-miosix.h	1970-01-01 01:00:00.000000000 +0100
+++ gcc-4.5.2/gcc/gthr-miosix.h	2011-05-30 10:35:06.617194051 +0200
@@ -0,0 +1,53 @@
+
+#ifndef GCC_GHTR_MIOSIX_H
+#define GCC_GHTR_MIOSIX_H
+
+#include <pthread.h>
+#include <unistd.h>
+
+//Note to self: gthr.h contains useful information
+//on how a gthr-xxx.h should look like
+
+#define __GTHREADS 1
+#define __GTHREAD_HAS_COND 1
+//#define __GTHREADS_CXX0X 1 //Not yet, missing __gthread_*_timed*
+
+//In Miosix, threads are always enabled, period.
+#define __gthread_active_p() 1
+
+typedef pthread_t       __gthread_t;
+typedef pthread_key_t   __gthread_key_t; //This actually is unimplemented
+typedef pthread_once_t  __gthread_once_t;
+typedef pthread_mutex_t __gthread_mutex_t;
+typedef pthread_mutex_t __gthread_recursive_mutex_t;
+typedef pthread_cond_t  __gthread_cond_t;
+typedef struct timespec __gthread_time_t;
+
+#define __GTHREAD_MUTEX_INIT           PTHREAD_MUTEX_INITIALIZER
+#define __GTHREAD_ONCE_INIT            PTHREAD_ONCE_INIT
+#define __GTHREAD_RECURSIVE_MUTEX_INIT PTHREAD_MUTEX_RECURSIVE_INITIALIZER_NP
+#define __GTHREAD_COND_INIT            PTHREAD_COND_INITIALIZER
+#define __GTHREAD_TIME_INIT            {0,0}
+
+#define __gthread_once                    pthread_once
+#define __gthread_mutex_destroy           pthread_mutex_destroy
+#define __gthread_mutex_lock              pthread_mutex_lock
+#define __gthread_mutex_trylock           pthread_mutex_trylock
+#define __gthread_mutex_unlock            pthread_mutex_unlock
+#define __gthread_recursive_mutex_lock    pthread_mutex_lock
+#define __gthread_recursive_mutex_trylock pthread_mutex_trylock
+#define __gthread_recursive_mutex_unlock  pthread_mutex_unlock
+#define __gthread_cond_broadcast          pthread_cond_broadcast
+#define __gthread_cond_wait               pthread_cond_wait
+#define __gthread_cond_wait_recursive     pthread_cond_wait
+
+//These actually aren't implemented in Miosix, so code trying to use these will
+//fail to link, and for now it's the "desired" behaviour (better than failing
+//at runtime, at least). They are used somewhere in libstdc++ too, but it can
+//be patched to remove those uses.
+#define __gthread_key_create              pthread_key_create
+#define __gthread_key_delete              pthread_key_delete
+#define __gthread_getspecific             pthread_getspecific
+#define __gthread_setspecific             pthread_setspecific
+
+#endif //GCC_GHTR_MIOSIX_H
diff -ruN gcc-4.5.2-original/gcc/unwind-sjlj.c gcc-4.5.2/gcc/unwind-sjlj.c
--- gcc-4.5.2-original/gcc/unwind-sjlj.c	2009-04-09 17:00:19.000000000 +0200
+++ gcc-4.5.2/gcc/unwind-sjlj.c	2011-05-30 10:35:06.618194025 +0200
@@ -94,74 +94,29 @@
 
 /* Manage the chain of registered function contexts.  */
 
-/* Single threaded fallback chain.  */
-static struct SjLj_Function_Context *fc_static;
-
-#if __GTHREADS
-static __gthread_key_t fc_key;
-static int use_fc_key = -1;
-
-static void
-fc_key_init (void)
-{
-  use_fc_key = __gthread_key_create (&fc_key, 0) == 0;
-}
-
-static void
-fc_key_init_once (void)
-{
-  static __gthread_once_t once = __GTHREAD_ONCE_INIT;
-  if (__gthread_once (&once, fc_key_init) != 0 || use_fc_key < 0)
-    use_fc_key = 0;
-}
-#endif
+//Begin TFT patch
+void _Miosix_set_sjlj_ptr(void* ptr);
+void *_Miosix_get_sjlj_ptr();
 
 void
 _Unwind_SjLj_Register (struct SjLj_Function_Context *fc)
 {
-#if __GTHREADS
-  if (use_fc_key < 0)
-    fc_key_init_once ();
-
-  if (use_fc_key)
-    {
-      fc->prev = __gthread_getspecific (fc_key);
-      __gthread_setspecific (fc_key, fc);
-    }
-  else
-#endif
-    {
-      fc->prev = fc_static;
-      fc_static = fc;
-    }
+  fc->prev=_Miosix_get_sjlj_ptr();
+  _Miosix_set_sjlj_ptr(fc);
 }
 
 static inline struct SjLj_Function_Context *
 _Unwind_SjLj_GetContext (void)
 {
-#if __GTHREADS
-  if (use_fc_key < 0)
-    fc_key_init_once ();
-
-  if (use_fc_key)
-    return __gthread_getspecific (fc_key);
-#endif
-  return fc_static;
+  return _Miosix_get_sjlj_ptr();
 }
 
 static inline void
 _Unwind_SjLj_SetContext (struct SjLj_Function_Context *fc)
 {
-#if __GTHREADS
-  if (use_fc_key < 0)
-    fc_key_init_once ();
-
-  if (use_fc_key)
-    __gthread_setspecific (fc_key, fc);
-  else
-#endif
-    fc_static = fc;
+  _Miosix_set_sjlj_ptr(fc);
 }
+//End TFT patch
 
 void
 _Unwind_SjLj_Unregister (struct SjLj_Function_Context *fc)
diff -ruN gcc-4.5.2-original/libstdc++-v3/config/cpu/arm/atomicity.h gcc-4.5.2/libstdc++-v3/config/cpu/arm/atomicity.h
--- gcc-4.5.2-original/libstdc++-v3/config/cpu/arm/atomicity.h	1970-01-01 01:00:00.000000000 +0100
+++ gcc-4.5.2/libstdc++-v3/config/cpu/arm/atomicity.h	2011-05-30 10:35:06.618194025 +0200
@@ -0,0 +1,36 @@
+//Begin TFT patch
+//Support _Atomic_word operations on arm-miosix-eabi
+#include <ext/atomicity.h>
+#include <ext/concurrence.h>
+
+//Declaring those here to avoid including miosix headers in libstdc++
+namespace miosix {
+void disableInterrupts();
+void enableInterrupts();
+}
+
+_GLIBCXX_BEGIN_NAMESPACE(__gnu_cxx)
+
+  _Atomic_word
+  __attribute__ ((__unused__))
+  __exchange_and_add(volatile _Atomic_word* __mem, int __val) throw ()
+  {
+    miosix::disableInterrupts();
+    _Atomic_word __result;
+    __result = *__mem;
+    *__mem += __val;
+    miosix::enableInterrupts();
+    return __result;
+  }
+
+  void
+  __attribute__ ((__unused__))
+  __atomic_add(volatile _Atomic_word* __mem, int __val) throw ()
+  {
+    miosix::disableInterrupts();
+    *__mem += __val;
+    miosix::enableInterrupts();
+  }
+
+_GLIBCXX_END_NAMESPACE
+//End TFT patch
diff -ruN gcc-4.5.2-original/libstdc++-v3/config/cpu/arm/atomic_word.h gcc-4.5.2/libstdc++-v3/config/cpu/arm/atomic_word.h
--- gcc-4.5.2-original/libstdc++-v3/config/cpu/arm/atomic_word.h	1970-01-01 01:00:00.000000000 +0100
+++ gcc-4.5.2/libstdc++-v3/config/cpu/arm/atomic_word.h	2011-05-30 10:35:06.618194025 +0200
@@ -0,0 +1,12 @@
+//begin TFT patch
+//Support _Atomic_word operations on arm-miosix-eabi
+#ifndef _GLIBCXX_ATOMIC_WORD_H
+#define _GLIBCXX_ATOMIC_WORD_H	1
+
+typedef int _Atomic_word;
+
+#define _GLIBCXX_READ_MEM_BARRIER __asm __volatile ("":::"memory")
+#define _GLIBCXX_WRITE_MEM_BARRIER __asm __volatile ("":::"memory")
+
+#endif 
+//end TFT patch
\ No newline at end of file
diff -ruN gcc-4.5.2-original/libstdc++-v3/include/Makefile.am gcc-4.5.2/libstdc++-v3/include/Makefile.am
--- gcc-4.5.2-original/libstdc++-v3/include/Makefile.am	2010-02-10 20:14:33.000000000 +0100
+++ gcc-4.5.2/libstdc++-v3/include/Makefile.am	2011-05-30 10:35:06.619194002 +0200
@@ -864,7 +864,8 @@
 	${host_builddir}/gthr-single.h \
 	${host_builddir}/gthr-posix.h \
 	${host_builddir}/gthr-tpf.h \
-	${host_builddir}/gthr-default.h
+	${host_builddir}/gthr-default.h \
+	${host_builddir}/gthr-miosix.h
 
 
 pch1_source = ${glibcxx_srcdir}/include/precompiled/stdc++.h
@@ -1153,6 +1154,14 @@
 	    -e 's,^#include "\(.*\)",#include <bits/\1>,g' \
 	    < ${toplevel_srcdir}/gcc/${glibcxx_thread_h} > $@
 
+${host_builddir}/gthr-miosix.h: ${toplevel_srcdir}/gcc/gthr-miosix.h \
+				   stamp-${host_alias}
+	sed -e 's/\(UNUSED\)/_GLIBCXX_\1/g' \
+	    -e 's/\(GCC${uppercase}*_H\)/_GLIBCXX_\1/g' \
+	    -e 's/SUPPORTS_WEAK/__GXX_WEAK__/g' \
+	    -e 's/\(${uppercase}*USE_WEAK\)/_GLIBCXX_\1/g' \
+	    < ${toplevel_srcdir}/gcc/gthr-miosix.h > $@
+
 # Build two precompiled C++ includes, stdc++.h.gch/*.gch
 ${pch1a_output}: ${allstamped} ${host_builddir}/c++config.h ${pch1_source}
 	-mkdir -p ${pch1_output_builddir}
diff -ruN gcc-4.5.2-original/libstdc++-v3/include/Makefile.in gcc-4.5.2/libstdc++-v3/include/Makefile.in
--- gcc-4.5.2-original/libstdc++-v3/include/Makefile.in	2010-04-02 20:18:06.000000000 +0200
+++ gcc-4.5.2/libstdc++-v3/include/Makefile.in	2011-05-30 10:35:06.620193984 +0200
@@ -1093,7 +1093,8 @@
 	${host_builddir}/gthr-single.h \
 	${host_builddir}/gthr-posix.h \
 	${host_builddir}/gthr-tpf.h \
-	${host_builddir}/gthr-default.h
+	${host_builddir}/gthr-default.h \
+	${host_builddir}/gthr-miosix.h
 
 pch1_source = ${glibcxx_srcdir}/include/precompiled/stdc++.h
 pch1_output_builddir = ${host_builddir}/stdc++.h.gch
@@ -1560,6 +1561,14 @@
 	    -e 's,^#include "\(.*\)",#include <bits/\1>,g' \
 	    < ${toplevel_srcdir}/gcc/${glibcxx_thread_h} > $@
 
+${host_builddir}/gthr-miosix.h: ${toplevel_srcdir}/gcc/gthr-miosix.h \
+				   stamp-${host_alias}
+	sed -e 's/\(UNUSED\)/_GLIBCXX_\1/g' \
+	    -e 's/\(GCC${uppercase}*_H\)/_GLIBCXX_\1/g' \
+	    -e 's/SUPPORTS_WEAK/__GXX_WEAK__/g' \
+	    -e 's/\(${uppercase}*USE_WEAK\)/_GLIBCXX_\1/g' \
+	    < ${toplevel_srcdir}/gcc/gthr-miosix.h > $@
+
 # Build two precompiled C++ includes, stdc++.h.gch/*.gch
 ${pch1a_output}: ${allstamped} ${host_builddir}/c++config.h ${pch1_source}
 	-mkdir -p ${pch1_output_builddir}
diff -ruN gcc-4.5.2-original/libstdc++-v3/libsupc++/eh_alloc.cc gcc-4.5.2/libstdc++-v3/libsupc++/eh_alloc.cc
--- gcc-4.5.2-original/libstdc++-v3/libsupc++/eh_alloc.cc	2009-04-09 17:00:19.000000000 +0200
+++ gcc-4.5.2/libstdc++-v3/libsupc++/eh_alloc.cc	2011-05-30 10:35:06.620193984 +0200
@@ -73,6 +73,15 @@
 # define EMERGENCY_OBJ_COUNT	4
 #endif
 
+//begin TFT patch
+//RATIONALE: reduced emergency buffer for ARM microcontrollers, saves ~1.5K of RAM
+#undef EMERGENCY_OBJ_SIZE
+#undef EMERGENCY_OBJ_COUNT
+
+#define EMERGENCY_OBJ_SIZE  160
+#define EMERGENCY_OBJ_COUNT 3
+//end TFT patch
+
 #if INT_MAX == 32767 || EMERGENCY_OBJ_COUNT <= 32
 typedef unsigned int bitmask_type;
 #else
diff -ruN gcc-4.5.2-original/libstdc++-v3/libsupc++/eh_globals.cc gcc-4.5.2/libstdc++-v3/libsupc++/eh_globals.cc
--- gcc-4.5.2-original/libstdc++-v3/libsupc++/eh_globals.cc	2009-04-09 17:00:19.000000000 +0200
+++ gcc-4.5.2/libstdc++-v3/libsupc++/eh_globals.cc	2011-05-30 10:35:06.621193968 +0200
@@ -42,119 +42,6 @@
 
 using namespace __cxxabiv1;
 
-#if _GLIBCXX_HAVE_TLS
-
-namespace
-{
-  abi::__cxa_eh_globals*
-  get_global() throw()
-  {
-    static __thread abi::__cxa_eh_globals global;
-    return &global;
-  }
-} // anonymous namespace
-
-extern "C" __cxa_eh_globals*
-__cxxabiv1::__cxa_get_globals_fast() throw()
-{ return get_global(); }
-
-extern "C" __cxa_eh_globals*
-__cxxabiv1::__cxa_get_globals() throw()
-{ return get_global(); }
-
-
-#else
-
-// Single-threaded fallback buffer.
-static __cxa_eh_globals eh_globals;
-
-#if __GTHREADS
-
-static void
-eh_globals_dtor(void* ptr)
-{
-  if (ptr)
-    {
-      __cxa_eh_globals* g = reinterpret_cast<__cxa_eh_globals*>(ptr);
-      __cxa_exception* exn = g->caughtExceptions;
-      __cxa_exception* next;
-      while (exn)
-	{
-	  next = exn->nextException;
-	  _Unwind_DeleteException(&exn->unwindHeader);
-	  exn = next;
-	}
-      free(ptr);
-    }
-}
-
-struct __eh_globals_init
-{
-  __gthread_key_t  	_M_key;
-  bool 			_M_init;
-
-  __eh_globals_init() : _M_init(false)
-  { 
-    if (__gthread_active_p())
-      _M_init = __gthread_key_create(&_M_key, eh_globals_dtor) == 0; 
-  }
-
-  ~__eh_globals_init()
-  {
-    if (_M_init)
-      __gthread_key_delete(_M_key);
-    _M_init = false;
-  }
-};
-
-static __eh_globals_init init;
-
-extern "C" __cxa_eh_globals*
-__cxxabiv1::__cxa_get_globals_fast() throw()
-{
-  __cxa_eh_globals* g;
-  if (init._M_init)
-    g = static_cast<__cxa_eh_globals*>(__gthread_getspecific(init._M_key));
-  else
-    g = &eh_globals;
-  return g;
-}
-
-extern "C" __cxa_eh_globals*
-__cxxabiv1::__cxa_get_globals() throw()
-{
-  __cxa_eh_globals* g;
-  if (init._M_init)
-    {
-      g = static_cast<__cxa_eh_globals*>(__gthread_getspecific(init._M_key));
-      if (!g)
-	{
-	  void* v = malloc(sizeof(__cxa_eh_globals));
-	  if (v == 0 || __gthread_setspecific(init._M_key, v) != 0)
-	    std::terminate();
-	  g = static_cast<__cxa_eh_globals*>(v);
-	  g->caughtExceptions = 0;
-	  g->uncaughtExceptions = 0;
-#ifdef __ARM_EABI_UNWINDER__
-	  g->propagatingExceptions = 0;
-#endif
-	}
-    }
-  else
-    g = &eh_globals;
-  return g;
-}
-
-#else
-
-extern "C" __cxa_eh_globals*
-__cxxabiv1::__cxa_get_globals_fast() throw()
-{ return &eh_globals; }
-
-extern "C" __cxa_eh_globals*
-__cxxabiv1::__cxa_get_globals() throw()
-{ return &eh_globals; }
-
-#endif
-
-#endif
+//Note by TFT: __cxa_get_globals() and __cxa_get_globals_fast()
+//are now Miosix syscalls, since the __cxa_eh_globals struct needs
+//to be provided on a per-thread basis
diff -ruN gcc-4.5.2-original/libstdc++-v3/libsupc++/guard.cc gcc-4.5.2/libstdc++-v3/libsupc++/guard.cc
--- gcc-4.5.2-original/libstdc++-v3/libsupc++/guard.cc	2009-11-09 23:09:30.000000000 +0100
+++ gcc-4.5.2/libstdc++-v3/libsupc++/guard.cc	2011-05-30 10:35:06.621193968 +0200
@@ -182,216 +182,5 @@
 //  | _GLIBCXX_GUARD_WAITING_BIT) and some other threads are waiting until
 //				  it is initialized.
 
-namespace __cxxabiv1 
-{
-#ifdef _GLIBCXX_USE_FUTEX
-  namespace
-  {
-    static inline int __guard_test_bit (const int __byte, const int __val)
-    {
-      union { int __i; char __c[sizeof (int)]; } __u = { 0 };
-      __u.__c[__byte] = __val;
-      return __u.__i;
-    }
-  }
-#endif
 
-  static inline int
-  init_in_progress_flag(__guard* g)
-  { return ((char *)g)[1]; }
-
-  static inline void
-  set_init_in_progress_flag(__guard* g, int v)
-  { ((char *)g)[1] = v; }
-
-  static inline void
-  throw_recursive_init_exception()
-  {
-#ifdef __EXCEPTIONS
-	throw __gnu_cxx::recursive_init_error();
-#else
-	// Use __builtin_trap so we don't require abort().
-	__builtin_trap();
-#endif
-  }
-
-  // acuire() is a helper function used to acquire guard if thread support is
-  // not compiled in or is compiled in but not enabled at run-time.
-  static int
-  acquire(__guard *g)
-  {
-    // Quit if the object is already initialized.
-    if (_GLIBCXX_GUARD_TEST(g))
-      return 0;
-
-    if (init_in_progress_flag(g))
-      throw_recursive_init_exception();
-
-    set_init_in_progress_flag(g, 1);
-    return 1;
-  }
-
-  extern "C"
-  int __cxa_guard_acquire (__guard *g) 
-  {
-#ifdef __GTHREADS
-    // If the target can reorder loads, we need to insert a read memory
-    // barrier so that accesses to the guarded variable happen after the
-    // guard test.
-    if (_GLIBCXX_GUARD_TEST_AND_ACQUIRE (g))
-      return 0;
-
-# ifdef _GLIBCXX_USE_FUTEX
-    // If __sync_* and futex syscall are supported, don't use any global
-    // mutex.
-    if (__gthread_active_p ())
-      {
-	int *gi = (int *) (void *) g;
-	const int guard_bit = _GLIBCXX_GUARD_BIT;
-	const int pending_bit = _GLIBCXX_GUARD_PENDING_BIT;
-	const int waiting_bit = _GLIBCXX_GUARD_WAITING_BIT;
-
-	while (1)
-	  {
-	    int old = __sync_val_compare_and_swap (gi, 0, pending_bit);
-	    if (old == 0)
-	      return 1;	// This thread should do the initialization.
-
-	    if (old == guard_bit)
-	      return 0;	// Already initialized.
-
-	    if (old == pending_bit)
-	      {
-		int newv = old | waiting_bit;
-		if (__sync_val_compare_and_swap (gi, old, newv) != old)
-		  continue;
-
-		old = newv;
-	      }
-
-	    syscall (SYS_futex, gi, _GLIBCXX_FUTEX_WAIT, old, 0);
-	  }
-      }
-# else
-    if (__gthread_active_p ())
-      {
-	mutex_wrapper mw;
-
-	while (1)	// When this loop is executing, mutex is locked.
-	  {
-#  ifdef __GTHREAD_HAS_COND
-	    // The static is already initialized.
-	    if (_GLIBCXX_GUARD_TEST(g))
-	      return 0;	// The mutex will be unlocked via wrapper
-
-	    if (init_in_progress_flag(g))
-	      {
-		// The guarded static is currently being initialized by
-		// another thread, so we release mutex and wait for the
-		// conditional variable. We will lock the mutex again after
-		// this.
-		get_static_cond().wait_recursive(&get_static_mutex());
-	      }
-	    else
-	      {
-		set_init_in_progress_flag(g, 1);
-		return 1; // The mutex will be unlocked via wrapper.
-	      }
-#  else
-	    // This provides compatibility with older systems not supporting
-	    // POSIX like conditional variables.
-	    if (acquire(g))
-	      {
-		mw.unlock = false;
-		return 1; // The mutex still locked.
-	      }
-	    return 0; // The mutex will be unlocked via wrapper.
-#  endif
-	  }
-      }
-# endif
-#endif
-
-    return acquire (g);
-  }
-
-  extern "C"
-  void __cxa_guard_abort (__guard *g) throw ()
-  {
-#ifdef _GLIBCXX_USE_FUTEX
-    // If __sync_* and futex syscall are supported, don't use any global
-    // mutex.
-    if (__gthread_active_p ())
-      {
-	int *gi = (int *) (void *) g;
-	const int waiting_bit = _GLIBCXX_GUARD_WAITING_BIT;
-	int old = __sync_lock_test_and_set (gi, 0);
-
-	if ((old & waiting_bit) != 0)
-	  syscall (SYS_futex, gi, _GLIBCXX_FUTEX_WAKE, INT_MAX);
-	return;
-      }
-#elif defined(__GTHREAD_HAS_COND)
-    if (__gthread_active_p())
-      {	
-	mutex_wrapper mw;
-
-	set_init_in_progress_flag(g, 0);
-
-	// If we abort, we still need to wake up all other threads waiting for
-	// the conditional variable.
-        get_static_cond().broadcast();
-	return;
-      }	
-#endif
-
-    set_init_in_progress_flag(g, 0);
-#if defined(__GTHREADS) && !defined(__GTHREAD_HAS_COND)
-    // This provides compatibility with older systems not supporting POSIX like
-    // conditional variables.
-    if (__gthread_active_p ())
-      static_mutex->unlock();
-#endif
-  }
-
-  extern "C"
-  void __cxa_guard_release (__guard *g) throw ()
-  {
-#ifdef _GLIBCXX_USE_FUTEX
-    // If __sync_* and futex syscall are supported, don't use any global
-    // mutex.
-    if (__gthread_active_p ())
-      {
-	int *gi = (int *) (void *) g;
-	const int guard_bit = _GLIBCXX_GUARD_BIT;
-	const int waiting_bit = _GLIBCXX_GUARD_WAITING_BIT;
-	int old = __sync_lock_test_and_set (gi, guard_bit);
-
-	if ((old & waiting_bit) != 0)
-	  syscall (SYS_futex, gi, _GLIBCXX_FUTEX_WAKE, INT_MAX);
-	return;
-      }
-#elif defined(__GTHREAD_HAS_COND)
-    if (__gthread_active_p())
-      {
-	mutex_wrapper mw;
-
-	set_init_in_progress_flag(g, 0);
-	_GLIBCXX_GUARD_SET_AND_RELEASE(g);
-
-        get_static_cond().broadcast();
-	return;
-      }	
-#endif
-
-    set_init_in_progress_flag(g, 0);
-    _GLIBCXX_GUARD_SET_AND_RELEASE (g);
-
-#if defined(__GTHREADS) && !defined(__GTHREAD_HAS_COND)
-    // This provides compatibility with older systems not supporting POSIX like
-    // conditional variables.
-    if (__gthread_active_p())
-      static_mutex->unlock();
-#endif
-  }
-}
+//Patched by TFT: __cxa_guard_[acquire|release|abort] made Miosix syscalls
diff -ruN gcc-4.5.2-original/libstdc++-v3/libsupc++/pure.cc gcc-4.5.2/libstdc++-v3/libsupc++/pure.cc
--- gcc-4.5.2-original/libstdc++-v3/libsupc++/pure.cc	2009-04-09 17:00:19.000000000 +0200
+++ gcc-4.5.2/libstdc++-v3/libsupc++/pure.cc	2011-05-31 14:44:00.957837835 +0200
@@ -43,8 +43,9 @@
 # define writestr(str) /* Empty */
 #endif
 
+//Patch by TFT: added __attribute__((weak))
 extern "C" void
-__cxxabiv1::__cxa_pure_virtual (void)
+__attribute__((weak)) __cxxabiv1::__cxa_pure_virtual (void)
 {
   writestr ("pure virtual method called\n");
   std::terminate ();
diff -ruN gcc-4.5.2-original/libstdc++-v3/libsupc++/vterminate.cc gcc-4.5.2/libstdc++-v3/libsupc++/vterminate.cc
--- gcc-4.5.2-original/libstdc++-v3/libsupc++/vterminate.cc	2009-09-29 02:15:30.000000000 +0200
+++ gcc-4.5.2/libstdc++-v3/libsupc++/vterminate.cc	2011-05-31 14:41:27.143837834 +0200
@@ -39,7 +39,9 @@
   // A replacement for the standard terminate_handler which prints
   // more information about the terminating exception (if any) on
   // stderr.
-  void __verbose_terminate_handler()
+
+  //Patch by TFT: add __attribute__((weak)) to make it overridable
+  void __attribute__((weak)) __verbose_terminate_handler()
   {
     static bool terminating;
     if (terminating)
diff -ruN gcc-4.5.2-original/libstdc++-v3/src/functexcept.cc gcc-4.5.2/libstdc++-v3/src/functexcept.cc
--- gcc-4.5.2-original/libstdc++-v3/src/functexcept.cc	2010-01-21 01:01:47.000000000 +0100
+++ gcc-4.5.2/libstdc++-v3/src/functexcept.cc	2011-05-31 14:46:55.800837835 +0200
@@ -40,72 +40,76 @@
 
 _GLIBCXX_BEGIN_NAMESPACE(std)
 
+//Patch by TFT: added __attribute__((weak)) to all these functions.
+//This to allow redefining them if compiling without exceptions to
+//avoid pulling in exception support (code size minimization)
+
 #if __EXCEPTIONS
-  void
+  void __attribute__((weak))
   __throw_bad_exception(void)
   { throw bad_exception(); }
 
-  void
+  void __attribute__((weak))
   __throw_bad_alloc(void)
   { throw bad_alloc(); }
 
-  void
+  void __attribute__((weak))
   __throw_bad_cast(void)
   { throw bad_cast(); }
 
-  void
+  void __attribute__((weak))
   __throw_bad_typeid(void)
   { throw bad_typeid(); }
 
-  void
+  void __attribute__((weak))
   __throw_logic_error(const char* __s)
   { throw logic_error(_(__s)); }
 
-  void
+  void __attribute__((weak))
   __throw_domain_error(const char* __s)
   { throw domain_error(_(__s)); }
 
-  void
+  void __attribute__((weak))
   __throw_invalid_argument(const char* __s)
   { throw invalid_argument(_(__s)); }
 
-  void
+  void __attribute__((weak))
   __throw_length_error(const char* __s)
   { throw length_error(_(__s)); }
 
-  void
+  void __attribute__((weak))
   __throw_out_of_range(const char* __s)
   { throw out_of_range(_(__s)); }
 
-  void
+  void __attribute__((weak))
   __throw_runtime_error(const char* __s)
   { throw runtime_error(_(__s)); }
 
-  void
+  void __attribute__((weak))
   __throw_range_error(const char* __s)
   { throw range_error(_(__s)); }
 
-  void
+  void __attribute__((weak))
   __throw_overflow_error(const char* __s)
   { throw overflow_error(_(__s)); }
 
-  void
+  void __attribute__((weak))
   __throw_underflow_error(const char* __s)
   { throw underflow_error(_(__s)); }
 
-  void
+  void __attribute__((weak))
   __throw_ios_failure(const char* __s)
   { throw ios_base::failure(_(__s)); }
 
-  void
+  void __attribute__((weak))
   __throw_system_error(int __i)
   { throw system_error(error_code(__i, generic_category())); }
 
-  void
+  void __attribute__((weak))
   __throw_future_error(int __i)
   { throw future_error(make_error_code(future_errc(__i))); }
 
-  void
+  void __attribute__((weak))
   __throw_bad_function_call()
   { throw bad_function_call(); }
 #else
