diff -ruN binutils-2.21.1/gas/testsuite/gas/arm/arch7.d binutils-2.21.1_patched/gas/testsuite/gas/arm/arch7.d
--- binutils-2.21.1/gas/testsuite/gas/arm/arch7.d	2010-09-17 12:42:04.000000000 +0200
+++ binutils-2.21.1_patched/gas/testsuite/gas/arm/arch7.d	2011-04-10 23:58:26.983682374 +0200
@@ -76,3 +76,4 @@
 0+10c <[^>]*> f380 8814 	msr	CONTROL, r0
 0+110 <[^>]*> f3ef 8003 	mrs	r0, PSR
 0+114 <[^>]*> f380 8803 	msr	PSR, r0
+0+118 <[^>]*> df00      	svc	0
diff -ruN binutils-2.21.1/gas/testsuite/gas/arm/arch7.s binutils-2.21.1_patched/gas/testsuite/gas/arm/arch7.s
--- binutils-2.21.1/gas/testsuite/gas/arm/arch7.s	2010-09-17 12:42:04.000000000 +0200
+++ binutils-2.21.1_patched/gas/testsuite/gas/arm/arch7.s	2011-04-10 23:59:35.216682373 +0200
@@ -79,3 +79,5 @@
 	msr	control, r0
 	mrs	r0, xpsr
 	msr	xpsr, r0
+
+	svc	0
\ No newline at end of file
diff -ruN binutils-2.21.1/gas/testsuite/gas/arm/attr-march-armv7.d binutils-2.21.1_patched/gas/testsuite/gas/arm/attr-march-armv7.d
--- binutils-2.21.1/gas/testsuite/gas/arm/attr-march-armv7.d	2010-05-11 19:36:33.000000000 +0200
+++ binutils-2.21.1_patched/gas/testsuite/gas/arm/attr-march-armv7.d	2011-04-11 00:01:04.542682372 +0200
@@ -9,5 +9,6 @@
 File Attributes
   Tag_CPU_name: "7"
   Tag_CPU_arch: v7
+  Tag_CPU_arch_profile: Microcontroller
   Tag_THUMB_ISA_use: Thumb-2
   Tag_DIV_use: Not allowed
diff -ruN binutils-2.21.1/include/opcode/arm.h binutils-2.21.1_patched/include/opcode/arm.h
--- binutils-2.21.1/include/opcode/arm.h	2010-11-15 11:03:05.000000000 +0100
+++ binutils-2.21.1_patched/include/opcode/arm.h	2011-04-11 00:02:45.094682373 +0200
@@ -109,7 +109,8 @@
 #define ARM_AEXT_V6KT2  (ARM_AEXT_V6T2 | ARM_EXT_V6K)
 #define ARM_AEXT_V6ZT2  (ARM_AEXT_V6T2 | ARM_EXT_SEC)
 #define ARM_AEXT_V6ZKT2 (ARM_AEXT_V6T2 | ARM_EXT_V6K | ARM_EXT_SEC)
-#define ARM_AEXT_V7_ARM	(ARM_AEXT_V6KT2 | ARM_EXT_V7 | ARM_EXT_BARRIER)
+#define ARM_AEXT_V7_ARM	(ARM_AEXT_V6KT2 | ARM_EXT_V7 | ARM_EXT_BARRIER \
+    | ARM_EXT_OS)
 #define ARM_AEXT_V7A	(ARM_AEXT_V7_ARM | ARM_EXT_V7A)
 #define ARM_AEXT_V7R	(ARM_AEXT_V7_ARM | ARM_EXT_V7R | ARM_EXT_DIV)
 #define ARM_AEXT_NOTM \
