diff -Naur Include.orig/stm32l412xx.h Include/stm32l412xx.h
--- Include.orig/stm32l412xx.h	2023-07-11 17:35:44
+++ Include/stm32l412xx.h	2024-01-06 20:37:25
@@ -137,7 +137,7 @@
   * @}
   */
 
-#include "core_cm4.h"             /* Cortex-M4 processor and core peripherals */
+#include <CMSIS/Include/core_cm4.h>             /* Cortex-M4 processor and core peripherals */
 #include "system_stm32l4xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l422xx.h Include/stm32l422xx.h
--- Include.orig/stm32l422xx.h	2023-07-11 17:35:44
+++ Include/stm32l422xx.h	2024-01-06 20:37:25
@@ -138,7 +138,7 @@
   * @}
   */
 
-#include "core_cm4.h"             /* Cortex-M4 processor and core peripherals */
+#include <CMSIS/Include/core_cm4.h>             /* Cortex-M4 processor and core peripherals */
 #include "system_stm32l4xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l431xx.h Include/stm32l431xx.h
--- Include.orig/stm32l431xx.h	2023-07-11 17:35:44
+++ Include/stm32l431xx.h	2024-01-06 20:37:25
@@ -145,7 +145,7 @@
   * @}
   */
 
-#include "core_cm4.h"             /* Cortex-M4 processor and core peripherals */
+#include <CMSIS/Include/core_cm4.h>             /* Cortex-M4 processor and core peripherals */
 #include "system_stm32l4xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l432xx.h Include/stm32l432xx.h
--- Include.orig/stm32l432xx.h	2023-07-11 17:35:44
+++ Include/stm32l432xx.h	2024-01-06 20:37:25
@@ -141,7 +141,7 @@
   * @}
   */
 
-#include "core_cm4.h"             /* Cortex-M4 processor and core peripherals */
+#include <CMSIS/Include/core_cm4.h>             /* Cortex-M4 processor and core peripherals */
 #include "system_stm32l4xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l433xx.h Include/stm32l433xx.h
--- Include.orig/stm32l433xx.h	2023-07-11 17:35:44
+++ Include/stm32l433xx.h	2024-01-06 20:37:25
@@ -147,7 +147,7 @@
   * @}
   */
 
-#include "core_cm4.h"             /* Cortex-M4 processor and core peripherals */
+#include <CMSIS/Include/core_cm4.h>             /* Cortex-M4 processor and core peripherals */
 #include "system_stm32l4xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l442xx.h Include/stm32l442xx.h
--- Include.orig/stm32l442xx.h	2023-07-11 17:35:44
+++ Include/stm32l442xx.h	2024-01-06 20:37:25
@@ -142,7 +142,7 @@
   * @}
   */
 
-#include "core_cm4.h"             /* Cortex-M4 processor and core peripherals */
+#include <CMSIS/Include/core_cm4.h>             /* Cortex-M4 processor and core peripherals */
 #include "system_stm32l4xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l443xx.h Include/stm32l443xx.h
--- Include.orig/stm32l443xx.h	2023-07-11 17:35:44
+++ Include/stm32l443xx.h	2024-01-06 20:37:25
@@ -148,7 +148,7 @@
   * @}
   */
 
-#include "core_cm4.h"             /* Cortex-M4 processor and core peripherals */
+#include <CMSIS/Include/core_cm4.h>             /* Cortex-M4 processor and core peripherals */
 #include "system_stm32l4xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l451xx.h Include/stm32l451xx.h
--- Include.orig/stm32l451xx.h	2023-07-11 17:35:44
+++ Include/stm32l451xx.h	2024-01-06 20:37:25
@@ -149,7 +149,7 @@
   * @}
   */
 
-#include "core_cm4.h"             /* Cortex-M4 processor and core peripherals */
+#include <CMSIS/Include/core_cm4.h>             /* Cortex-M4 processor and core peripherals */
 #include "system_stm32l4xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l452xx.h Include/stm32l452xx.h
--- Include.orig/stm32l452xx.h	2023-07-11 17:35:44
+++ Include/stm32l452xx.h	2024-01-06 20:37:25
@@ -150,7 +150,7 @@
   * @}
   */
 
-#include "core_cm4.h"             /* Cortex-M4 processor and core peripherals */
+#include <CMSIS/Include/core_cm4.h>             /* Cortex-M4 processor and core peripherals */
 #include "system_stm32l4xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l462xx.h Include/stm32l462xx.h
--- Include.orig/stm32l462xx.h	2023-07-11 17:35:44
+++ Include/stm32l462xx.h	2024-01-06 20:37:25
@@ -151,7 +151,7 @@
   * @}
   */
 
-#include "core_cm4.h"             /* Cortex-M4 processor and core peripherals */
+#include <CMSIS/Include/core_cm4.h>             /* Cortex-M4 processor and core peripherals */
 #include "system_stm32l4xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l471xx.h Include/stm32l471xx.h
--- Include.orig/stm32l471xx.h	2023-07-11 17:35:44
+++ Include/stm32l471xx.h	2024-01-06 20:37:25
@@ -160,7 +160,7 @@
   * @}
   */
 
-#include "core_cm4.h"             /* Cortex-M4 processor and core peripherals */
+#include <CMSIS/Include/core_cm4.h>             /* Cortex-M4 processor and core peripherals */
 #include "system_stm32l4xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l475xx.h Include/stm32l475xx.h
--- Include.orig/stm32l475xx.h	2023-07-11 17:35:44
+++ Include/stm32l475xx.h	2024-01-06 20:37:25
@@ -161,7 +161,7 @@
   * @}
   */
 
-#include "core_cm4.h"             /* Cortex-M4 processor and core peripherals */
+#include <CMSIS/Include/core_cm4.h>             /* Cortex-M4 processor and core peripherals */
 #include "system_stm32l4xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l476xx.h Include/stm32l476xx.h
--- Include.orig/stm32l476xx.h	2023-07-11 17:35:44
+++ Include/stm32l476xx.h	2024-01-06 20:37:25
@@ -162,7 +162,7 @@
   * @}
   */
 
-#include "core_cm4.h"             /* Cortex-M4 processor and core peripherals */
+#include <CMSIS/Include/core_cm4.h>             /* Cortex-M4 processor and core peripherals */
 #include "system_stm32l4xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l485xx.h Include/stm32l485xx.h
--- Include.orig/stm32l485xx.h	2023-07-11 17:35:44
+++ Include/stm32l485xx.h	2024-01-06 20:37:25
@@ -162,7 +162,7 @@
   * @}
   */
 
-#include "core_cm4.h"             /* Cortex-M4 processor and core peripherals */
+#include <CMSIS/Include/core_cm4.h>             /* Cortex-M4 processor and core peripherals */
 #include "system_stm32l4xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l486xx.h Include/stm32l486xx.h
--- Include.orig/stm32l486xx.h	2023-07-11 17:35:44
+++ Include/stm32l486xx.h	2024-01-06 20:37:25
@@ -163,7 +163,7 @@
   * @}
   */
 
-#include "core_cm4.h"             /* Cortex-M4 processor and core peripherals */
+#include <CMSIS/Include/core_cm4.h>             /* Cortex-M4 processor and core peripherals */
 #include "system_stm32l4xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l496xx.h Include/stm32l496xx.h
--- Include.orig/stm32l496xx.h	2023-07-11 17:35:44
+++ Include/stm32l496xx.h	2024-01-06 20:37:25
@@ -171,7 +171,7 @@
   * @}
   */
 
-#include "core_cm4.h"             /* Cortex-M4 processor and core peripherals */
+#include <CMSIS/Include/core_cm4.h>             /* Cortex-M4 processor and core peripherals */
 #include "system_stm32l4xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l4a6xx.h Include/stm32l4a6xx.h
--- Include.orig/stm32l4a6xx.h	2023-07-11 17:35:44
+++ Include/stm32l4a6xx.h	2024-01-06 20:37:25
@@ -172,7 +172,7 @@
   * @}
   */
 
-#include "core_cm4.h"             /* Cortex-M4 processor and core peripherals */
+#include <CMSIS/Include/core_cm4.h>             /* Cortex-M4 processor and core peripherals */
 #include "system_stm32l4xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l4p5xx.h Include/stm32l4p5xx.h
--- Include.orig/stm32l4p5xx.h	2023-07-11 17:35:44
+++ Include/stm32l4p5xx.h	2024-01-06 20:50:59
@@ -167,7 +167,7 @@
   * @}
   */
 
-#include "core_cm4.h"             /* Cortex-M4 processor and core peripherals */
+#include <CMSIS/Include/core_cm4.h>             /* Cortex-M4 processor and core peripherals */
 #include "system_stm32l4xx.h"
 #include <stdint.h>
 
@@ -18599,6 +18599,7 @@
 #define USART_CR1_RXNEIE_RXFNEIE_Pos   (5U)
 #define USART_CR1_RXNEIE_RXFNEIE_Msk   (0x1UL << USART_CR1_RXNEIE_RXFNEIE_Pos) /*!< 0x00000020 */
 #define USART_CR1_RXNEIE_RXFNEIE       USART_CR1_RXNEIE_RXFNEIE_Msk            /*!< RXNE/RXFIFO not empty Interrupt Enable */
+#define USART_CR1_RXNEIE              USART_CR1_RXNEIE_RXFNEIE                 /*!< BUGFIX Equivalent declaration for L4 and L4+ serial driver compatibility*/
 #define USART_CR1_TCIE_Pos            (6U)
 #define USART_CR1_TCIE_Msk            (0x1UL << USART_CR1_TCIE_Pos)            /*!< 0x00000040 */
 #define USART_CR1_TCIE                USART_CR1_TCIE_Msk                       /*!< Transmission Complete Interrupt Enable */
@@ -18878,12 +18879,14 @@
 #define USART_ISR_RXNE_RXFNE_Pos      (5U)
 #define USART_ISR_RXNE_RXFNE_Msk      (0x1UL << USART_ISR_RXNE_RXFNE_Pos)      /*!< 0x00000020 */
 #define USART_ISR_RXNE_RXFNE           USART_ISR_RXNE_RXFNE_Msk                /*!< Read Data Register Not Empty/RXFIFO Not Empty */
+#define USART_ISR_RXNE                 USART_ISR_RXNE_RXFNE                    /*!< BUGFIX Equivalent declaration for L4 and L4+ serial driver compatibility*/
 #define USART_ISR_TC_Pos              (6U)
 #define USART_ISR_TC_Msk              (0x1UL << USART_ISR_TC_Pos)              /*!< 0x00000040 */
 #define USART_ISR_TC                  USART_ISR_TC_Msk                         /*!< Transmission Complete */
 #define USART_ISR_TXE_TXFNF_Pos       (7U)
 #define USART_ISR_TXE_TXFNF_Msk       (0x1UL << USART_ISR_TXE_TXFNF_Pos)       /*!< 0x00000080 */
 #define USART_ISR_TXE_TXFNF           USART_ISR_TXE_TXFNF_Msk                  /*!< Transmit Data Register Empty/TXFIFO Not Full */
+#define USART_ISR_TXE                 USART_ISR_TXE_TXFNF                      /*!< BUGFIX Equivalent declaration for L4 and L4+ serial driver compatibility*/
 #define USART_ISR_LBDF_Pos            (8U)
 #define USART_ISR_LBDF_Msk            (0x1UL << USART_ISR_LBDF_Pos)            /*!< 0x00000100 */
 #define USART_ISR_LBDF                USART_ISR_LBDF_Msk                       /*!< LIN Break Detection Flag */
diff -Naur Include.orig/stm32l4q5xx.h Include/stm32l4q5xx.h
--- Include.orig/stm32l4q5xx.h	2023-07-11 17:35:44
+++ Include/stm32l4q5xx.h	2024-01-06 20:50:59
@@ -169,7 +169,7 @@
   * @}
   */
 
-#include "core_cm4.h"             /* Cortex-M4 processor and core peripherals */
+#include <CMSIS/Include/core_cm4.h>             /* Cortex-M4 processor and core peripherals */
 #include "system_stm32l4xx.h"
 #include <stdint.h>
 
@@ -19110,6 +19110,7 @@
 #define USART_CR1_RXNEIE_RXFNEIE_Pos   (5U)
 #define USART_CR1_RXNEIE_RXFNEIE_Msk   (0x1UL << USART_CR1_RXNEIE_RXFNEIE_Pos) /*!< 0x00000020 */
 #define USART_CR1_RXNEIE_RXFNEIE       USART_CR1_RXNEIE_RXFNEIE_Msk            /*!< RXNE/RXFIFO not empty Interrupt Enable */
+#define USART_CR1_RXNEIE              USART_CR1_RXNEIE_RXFNEIE                 /*!< BUGFIX Equivalent declaration for L4 and L4+ serial driver compatibility*/
 #define USART_CR1_TCIE_Pos            (6U)
 #define USART_CR1_TCIE_Msk            (0x1UL << USART_CR1_TCIE_Pos)            /*!< 0x00000040 */
 #define USART_CR1_TCIE                USART_CR1_TCIE_Msk                       /*!< Transmission Complete Interrupt Enable */
@@ -19389,12 +19390,14 @@
 #define USART_ISR_RXNE_RXFNE_Pos      (5U)
 #define USART_ISR_RXNE_RXFNE_Msk      (0x1UL << USART_ISR_RXNE_RXFNE_Pos)      /*!< 0x00000020 */
 #define USART_ISR_RXNE_RXFNE           USART_ISR_RXNE_RXFNE_Msk                /*!< Read Data Register Not Empty/RXFIFO Not Empty */
+#define USART_ISR_RXNE                 USART_ISR_RXNE_RXFNE                    /*!< BUGFIX Equivalent declaration for L4 and L4+ serial driver compatibility*/
 #define USART_ISR_TC_Pos              (6U)
 #define USART_ISR_TC_Msk              (0x1UL << USART_ISR_TC_Pos)              /*!< 0x00000040 */
 #define USART_ISR_TC                  USART_ISR_TC_Msk                         /*!< Transmission Complete */
 #define USART_ISR_TXE_TXFNF_Pos       (7U)
 #define USART_ISR_TXE_TXFNF_Msk       (0x1UL << USART_ISR_TXE_TXFNF_Pos)       /*!< 0x00000080 */
 #define USART_ISR_TXE_TXFNF           USART_ISR_TXE_TXFNF_Msk                  /*!< Transmit Data Register Empty/TXFIFO Not Full */
+#define USART_ISR_TXE                 USART_ISR_TXE_TXFNF                      /*!< BUGFIX Equivalent declaration for L4 and L4+ serial driver compatibility*/
 #define USART_ISR_LBDF_Pos            (8U)
 #define USART_ISR_LBDF_Msk            (0x1UL << USART_ISR_LBDF_Pos)            /*!< 0x00000100 */
 #define USART_ISR_LBDF                USART_ISR_LBDF_Msk                       /*!< LIN Break Detection Flag */
diff -Naur Include.orig/stm32l4r5xx.h Include/stm32l4r5xx.h
--- Include.orig/stm32l4r5xx.h	2023-07-11 17:35:44
+++ Include/stm32l4r5xx.h	2024-01-06 20:50:59
@@ -166,7 +166,7 @@
   * @}
   */
 
-#include "core_cm4.h"             /* Cortex-M4 processor and core peripherals */
+#include <CMSIS/Include/core_cm4.h>             /* Cortex-M4 processor and core peripherals */
 #include "system_stm32l4xx.h"
 #include <stdint.h>
 
@@ -17562,6 +17562,7 @@
 #define USART_CR1_RXNEIE_RXFNEIE_Pos   (5U)
 #define USART_CR1_RXNEIE_RXFNEIE_Msk   (0x1UL << USART_CR1_RXNEIE_RXFNEIE_Pos) /*!< 0x00000020 */
 #define USART_CR1_RXNEIE_RXFNEIE       USART_CR1_RXNEIE_RXFNEIE_Msk            /*!< RXNE/RXFIFO not empty Interrupt Enable */
+#define USART_CR1_RXNEIE              USART_CR1_RXNEIE_RXFNEIE                 /*!< BUGFIX Equivalent declaration for L4 and L4+ serial driver compatibility*/
 #define USART_CR1_TCIE_Pos            (6U)
 #define USART_CR1_TCIE_Msk            (0x1UL << USART_CR1_TCIE_Pos)            /*!< 0x00000040 */
 #define USART_CR1_TCIE                USART_CR1_TCIE_Msk                       /*!< Transmission Complete Interrupt Enable */
@@ -17841,12 +17842,14 @@
 #define USART_ISR_RXNE_RXFNE_Pos      (5U)
 #define USART_ISR_RXNE_RXFNE_Msk      (0x1UL << USART_ISR_RXNE_RXFNE_Pos)      /*!< 0x00000020 */
 #define USART_ISR_RXNE_RXFNE           USART_ISR_RXNE_RXFNE_Msk                /*!< Read Data Register Not Empty/RXFIFO Not Empty */
+#define USART_ISR_RXNE                 USART_ISR_RXNE_RXFNE                    /*!< BUGFIX Equivalent declaration for L4 and L4+ serial driver compatibility*/
 #define USART_ISR_TC_Pos              (6U)
 #define USART_ISR_TC_Msk              (0x1UL << USART_ISR_TC_Pos)              /*!< 0x00000040 */
 #define USART_ISR_TC                  USART_ISR_TC_Msk                         /*!< Transmission Complete */
 #define USART_ISR_TXE_TXFNF_Pos       (7U)
 #define USART_ISR_TXE_TXFNF_Msk       (0x1UL << USART_ISR_TXE_TXFNF_Pos)       /*!< 0x00000080 */
 #define USART_ISR_TXE_TXFNF           USART_ISR_TXE_TXFNF_Msk                  /*!< Transmit Data Register Empty/TXFIFO Not Full */
+#define USART_ISR_TXE                 USART_ISR_TXE_TXFNF                      /*!< BUGFIX Equivalent declaration for L4 and L4+ serial driver compatibility*/
 #define USART_ISR_LBDF_Pos            (8U)
 #define USART_ISR_LBDF_Msk            (0x1UL << USART_ISR_LBDF_Pos)            /*!< 0x00000100 */
 #define USART_ISR_LBDF                USART_ISR_LBDF_Msk                       /*!< LIN Break Detection Flag */
diff -Naur Include.orig/stm32l4r7xx.h Include/stm32l4r7xx.h
--- Include.orig/stm32l4r7xx.h	2023-07-11 17:35:44
+++ Include/stm32l4r7xx.h	2024-01-06 20:50:59
@@ -169,7 +169,7 @@
   * @}
   */
 
-#include "core_cm4.h"             /* Cortex-M4 processor and core peripherals */
+#include <CMSIS/Include/core_cm4.h>             /* Cortex-M4 processor and core peripherals */
 #include "system_stm32l4xx.h"
 #include <stdint.h>
 
@@ -18061,6 +18061,7 @@
 #define USART_CR1_RXNEIE_RXFNEIE_Pos   (5U)
 #define USART_CR1_RXNEIE_RXFNEIE_Msk   (0x1UL << USART_CR1_RXNEIE_RXFNEIE_Pos) /*!< 0x00000020 */
 #define USART_CR1_RXNEIE_RXFNEIE       USART_CR1_RXNEIE_RXFNEIE_Msk            /*!< RXNE/RXFIFO not empty Interrupt Enable */
+#define USART_CR1_RXNEIE              USART_CR1_RXNEIE_RXFNEIE                 /*!< BUGFIX Equivalent declaration for L4 and L4+ serial driver compatibility*/
 #define USART_CR1_TCIE_Pos            (6U)
 #define USART_CR1_TCIE_Msk            (0x1UL << USART_CR1_TCIE_Pos)            /*!< 0x00000040 */
 #define USART_CR1_TCIE                USART_CR1_TCIE_Msk                       /*!< Transmission Complete Interrupt Enable */
@@ -18340,12 +18341,14 @@
 #define USART_ISR_RXNE_RXFNE_Pos      (5U)
 #define USART_ISR_RXNE_RXFNE_Msk      (0x1UL << USART_ISR_RXNE_RXFNE_Pos)      /*!< 0x00000020 */
 #define USART_ISR_RXNE_RXFNE           USART_ISR_RXNE_RXFNE_Msk                /*!< Read Data Register Not Empty/RXFIFO Not Empty */
+#define USART_ISR_RXNE                 USART_ISR_RXNE_RXFNE                    /*!< BUGFIX Equivalent declaration for L4 and L4+ serial driver compatibility*/
 #define USART_ISR_TC_Pos              (6U)
 #define USART_ISR_TC_Msk              (0x1UL << USART_ISR_TC_Pos)              /*!< 0x00000040 */
 #define USART_ISR_TC                  USART_ISR_TC_Msk                         /*!< Transmission Complete */
 #define USART_ISR_TXE_TXFNF_Pos       (7U)
 #define USART_ISR_TXE_TXFNF_Msk       (0x1UL << USART_ISR_TXE_TXFNF_Pos)       /*!< 0x00000080 */
 #define USART_ISR_TXE_TXFNF           USART_ISR_TXE_TXFNF_Msk                  /*!< Transmit Data Register Empty/TXFIFO Not Full */
+#define USART_ISR_TXE                 USART_ISR_TXE_TXFNF                      /*!< BUGFIX Equivalent declaration for L4 and L4+ serial driver compatibility*/
 #define USART_ISR_LBDF_Pos            (8U)
 #define USART_ISR_LBDF_Msk            (0x1UL << USART_ISR_LBDF_Pos)            /*!< 0x00000100 */
 #define USART_ISR_LBDF                USART_ISR_LBDF_Msk                       /*!< LIN Break Detection Flag */
diff -Naur Include.orig/stm32l4r9xx.h Include/stm32l4r9xx.h
--- Include.orig/stm32l4r9xx.h	2023-07-11 17:35:44
+++ Include/stm32l4r9xx.h	2024-01-06 20:50:59
@@ -170,7 +170,7 @@
   * @}
   */
 
-#include "core_cm4.h"             /* Cortex-M4 processor and core peripherals */
+#include <CMSIS/Include/core_cm4.h>             /* Cortex-M4 processor and core peripherals */
 #include "system_stm32l4xx.h"
 #include <stdint.h>
 
@@ -21193,6 +21193,7 @@
 #define USART_CR1_RXNEIE_RXFNEIE_Pos   (5U)
 #define USART_CR1_RXNEIE_RXFNEIE_Msk   (0x1UL << USART_CR1_RXNEIE_RXFNEIE_Pos) /*!< 0x00000020 */
 #define USART_CR1_RXNEIE_RXFNEIE       USART_CR1_RXNEIE_RXFNEIE_Msk            /*!< RXNE/RXFIFO not empty Interrupt Enable */
+#define USART_CR1_RXNEIE              USART_CR1_RXNEIE_RXFNEIE                 /*!< BUGFIX Equivalent declaration for L4 and L4+ serial driver compatibility*/
 #define USART_CR1_TCIE_Pos            (6U)
 #define USART_CR1_TCIE_Msk            (0x1UL << USART_CR1_TCIE_Pos)            /*!< 0x00000040 */
 #define USART_CR1_TCIE                USART_CR1_TCIE_Msk                       /*!< Transmission Complete Interrupt Enable */
@@ -21472,12 +21473,14 @@
 #define USART_ISR_RXNE_RXFNE_Pos      (5U)
 #define USART_ISR_RXNE_RXFNE_Msk      (0x1UL << USART_ISR_RXNE_RXFNE_Pos)      /*!< 0x00000020 */
 #define USART_ISR_RXNE_RXFNE           USART_ISR_RXNE_RXFNE_Msk                /*!< Read Data Register Not Empty/RXFIFO Not Empty */
+#define USART_ISR_RXNE                 USART_ISR_RXNE_RXFNE                    /*!< BUGFIX Equivalent declaration for L4 and L4+ serial driver compatibility*/
 #define USART_ISR_TC_Pos              (6U)
 #define USART_ISR_TC_Msk              (0x1UL << USART_ISR_TC_Pos)              /*!< 0x00000040 */
 #define USART_ISR_TC                  USART_ISR_TC_Msk                         /*!< Transmission Complete */
 #define USART_ISR_TXE_TXFNF_Pos       (7U)
 #define USART_ISR_TXE_TXFNF_Msk       (0x1UL << USART_ISR_TXE_TXFNF_Pos)       /*!< 0x00000080 */
 #define USART_ISR_TXE_TXFNF           USART_ISR_TXE_TXFNF_Msk                  /*!< Transmit Data Register Empty/TXFIFO Not Full */
+#define USART_ISR_TXE                 USART_ISR_TXE_TXFNF                      /*!< BUGFIX Equivalent declaration for L4 and L4+ serial driver compatibility*/
 #define USART_ISR_LBDF_Pos            (8U)
 #define USART_ISR_LBDF_Msk            (0x1UL << USART_ISR_LBDF_Pos)            /*!< 0x00000100 */
 #define USART_ISR_LBDF                USART_ISR_LBDF_Msk                       /*!< LIN Break Detection Flag */
diff -Naur Include.orig/stm32l4s5xx.h Include/stm32l4s5xx.h
--- Include.orig/stm32l4s5xx.h	2023-07-11 17:35:44
+++ Include/stm32l4s5xx.h	2024-01-06 20:50:59
@@ -167,7 +167,7 @@
   * @}
   */
 
-#include "core_cm4.h"             /* Cortex-M4 processor and core peripherals */
+#include <CMSIS/Include/core_cm4.h>             /* Cortex-M4 processor and core peripherals */
 #include "system_stm32l4xx.h"
 #include <stdint.h>
 
@@ -17909,6 +17909,7 @@
 #define USART_CR1_RXNEIE_RXFNEIE_Pos   (5U)
 #define USART_CR1_RXNEIE_RXFNEIE_Msk   (0x1UL << USART_CR1_RXNEIE_RXFNEIE_Pos) /*!< 0x00000020 */
 #define USART_CR1_RXNEIE_RXFNEIE       USART_CR1_RXNEIE_RXFNEIE_Msk            /*!< RXNE/RXFIFO not empty Interrupt Enable */
+#define USART_CR1_RXNEIE              USART_CR1_RXNEIE_RXFNEIE                 /*!< BUGFIX Equivalent declaration for L4 and L4+ serial driver compatibility*/
 #define USART_CR1_TCIE_Pos            (6U)
 #define USART_CR1_TCIE_Msk            (0x1UL << USART_CR1_TCIE_Pos)            /*!< 0x00000040 */
 #define USART_CR1_TCIE                USART_CR1_TCIE_Msk                       /*!< Transmission Complete Interrupt Enable */
@@ -18188,12 +18189,14 @@
 #define USART_ISR_RXNE_RXFNE_Pos      (5U)
 #define USART_ISR_RXNE_RXFNE_Msk      (0x1UL << USART_ISR_RXNE_RXFNE_Pos)      /*!< 0x00000020 */
 #define USART_ISR_RXNE_RXFNE           USART_ISR_RXNE_RXFNE_Msk                /*!< Read Data Register Not Empty/RXFIFO Not Empty */
+#define USART_ISR_RXNE                 USART_ISR_RXNE_RXFNE                    /*!< BUGFIX Equivalent declaration for L4 and L4+ serial driver compatibility*/
 #define USART_ISR_TC_Pos              (6U)
 #define USART_ISR_TC_Msk              (0x1UL << USART_ISR_TC_Pos)              /*!< 0x00000040 */
 #define USART_ISR_TC                  USART_ISR_TC_Msk                         /*!< Transmission Complete */
 #define USART_ISR_TXE_TXFNF_Pos       (7U)
 #define USART_ISR_TXE_TXFNF_Msk       (0x1UL << USART_ISR_TXE_TXFNF_Pos)       /*!< 0x00000080 */
 #define USART_ISR_TXE_TXFNF           USART_ISR_TXE_TXFNF_Msk                  /*!< Transmit Data Register Empty/TXFIFO Not Full */
+#define USART_ISR_TXE                 USART_ISR_TXE_TXFNF                      /*!< BUGFIX Equivalent declaration for L4 and L4+ serial driver compatibility*/
 #define USART_ISR_LBDF_Pos            (8U)
 #define USART_ISR_LBDF_Msk            (0x1UL << USART_ISR_LBDF_Pos)            /*!< 0x00000100 */
 #define USART_ISR_LBDF                USART_ISR_LBDF_Msk                       /*!< LIN Break Detection Flag */
diff -Naur Include.orig/stm32l4s7xx.h Include/stm32l4s7xx.h
--- Include.orig/stm32l4s7xx.h	2023-07-11 17:35:44
+++ Include/stm32l4s7xx.h	2024-01-06 20:50:59
@@ -170,7 +170,7 @@
   * @}
   */
 
-#include "core_cm4.h"             /* Cortex-M4 processor and core peripherals */
+#include <CMSIS/Include/core_cm4.h>             /* Cortex-M4 processor and core peripherals */
 #include "system_stm32l4xx.h"
 #include <stdint.h>
 
@@ -18408,6 +18408,7 @@
 #define USART_CR1_RXNEIE_RXFNEIE_Pos   (5U)
 #define USART_CR1_RXNEIE_RXFNEIE_Msk   (0x1UL << USART_CR1_RXNEIE_RXFNEIE_Pos) /*!< 0x00000020 */
 #define USART_CR1_RXNEIE_RXFNEIE       USART_CR1_RXNEIE_RXFNEIE_Msk            /*!< RXNE/RXFIFO not empty Interrupt Enable */
+#define USART_CR1_RXNEIE              USART_CR1_RXNEIE_RXFNEIE                 /*!< BUGFIX Equivalent declaration for L4 and L4+ serial driver compatibility*/
 #define USART_CR1_TCIE_Pos            (6U)
 #define USART_CR1_TCIE_Msk            (0x1UL << USART_CR1_TCIE_Pos)            /*!< 0x00000040 */
 #define USART_CR1_TCIE                USART_CR1_TCIE_Msk                       /*!< Transmission Complete Interrupt Enable */
@@ -18687,12 +18688,14 @@
 #define USART_ISR_RXNE_RXFNE_Pos      (5U)
 #define USART_ISR_RXNE_RXFNE_Msk      (0x1UL << USART_ISR_RXNE_RXFNE_Pos)      /*!< 0x00000020 */
 #define USART_ISR_RXNE_RXFNE           USART_ISR_RXNE_RXFNE_Msk                /*!< Read Data Register Not Empty/RXFIFO Not Empty */
+#define USART_ISR_RXNE                 USART_ISR_RXNE_RXFNE                    /*!< BUGFIX Equivalent declaration for L4 and L4+ serial driver compatibility*/
 #define USART_ISR_TC_Pos              (6U)
 #define USART_ISR_TC_Msk              (0x1UL << USART_ISR_TC_Pos)              /*!< 0x00000040 */
 #define USART_ISR_TC                  USART_ISR_TC_Msk                         /*!< Transmission Complete */
 #define USART_ISR_TXE_TXFNF_Pos       (7U)
 #define USART_ISR_TXE_TXFNF_Msk       (0x1UL << USART_ISR_TXE_TXFNF_Pos)       /*!< 0x00000080 */
 #define USART_ISR_TXE_TXFNF           USART_ISR_TXE_TXFNF_Msk                  /*!< Transmit Data Register Empty/TXFIFO Not Full */
+#define USART_ISR_TXE                 USART_ISR_TXE_TXFNF                      /*!< BUGFIX Equivalent declaration for L4 and L4+ serial driver compatibility*/
 #define USART_ISR_LBDF_Pos            (8U)
 #define USART_ISR_LBDF_Msk            (0x1UL << USART_ISR_LBDF_Pos)            /*!< 0x00000100 */
 #define USART_ISR_LBDF                USART_ISR_LBDF_Msk                       /*!< LIN Break Detection Flag */
diff -Naur Include.orig/stm32l4s9xx.h Include/stm32l4s9xx.h
--- Include.orig/stm32l4s9xx.h	2023-07-11 17:35:44
+++ Include/stm32l4s9xx.h	2024-01-06 20:50:59
@@ -171,7 +171,7 @@
   * @}
   */
 
-#include "core_cm4.h"             /* Cortex-M4 processor and core peripherals */
+#include <CMSIS/Include/core_cm4.h>             /* Cortex-M4 processor and core peripherals */
 #include "system_stm32l4xx.h"
 #include <stdint.h>
 
@@ -21540,6 +21540,7 @@
 #define USART_CR1_RXNEIE_RXFNEIE_Pos   (5U)
 #define USART_CR1_RXNEIE_RXFNEIE_Msk   (0x1UL << USART_CR1_RXNEIE_RXFNEIE_Pos) /*!< 0x00000020 */
 #define USART_CR1_RXNEIE_RXFNEIE       USART_CR1_RXNEIE_RXFNEIE_Msk            /*!< RXNE/RXFIFO not empty Interrupt Enable */
+#define USART_CR1_RXNEIE              USART_CR1_RXNEIE_RXFNEIE                 /*!< BUGFIX Equivalent declaration for L4 and L4+ serial driver compatibility*/
 #define USART_CR1_TCIE_Pos            (6U)
 #define USART_CR1_TCIE_Msk            (0x1UL << USART_CR1_TCIE_Pos)            /*!< 0x00000040 */
 #define USART_CR1_TCIE                USART_CR1_TCIE_Msk                       /*!< Transmission Complete Interrupt Enable */
@@ -21819,12 +21820,14 @@
 #define USART_ISR_RXNE_RXFNE_Pos      (5U)
 #define USART_ISR_RXNE_RXFNE_Msk      (0x1UL << USART_ISR_RXNE_RXFNE_Pos)      /*!< 0x00000020 */
 #define USART_ISR_RXNE_RXFNE           USART_ISR_RXNE_RXFNE_Msk                /*!< Read Data Register Not Empty/RXFIFO Not Empty */
+#define USART_ISR_RXNE                 USART_ISR_RXNE_RXFNE                    /*!< BUGFIX Equivalent declaration for L4 and L4+ serial driver compatibility*/
 #define USART_ISR_TC_Pos              (6U)
 #define USART_ISR_TC_Msk              (0x1UL << USART_ISR_TC_Pos)              /*!< 0x00000040 */
 #define USART_ISR_TC                  USART_ISR_TC_Msk                         /*!< Transmission Complete */
 #define USART_ISR_TXE_TXFNF_Pos       (7U)
 #define USART_ISR_TXE_TXFNF_Msk       (0x1UL << USART_ISR_TXE_TXFNF_Pos)       /*!< 0x00000080 */
 #define USART_ISR_TXE_TXFNF           USART_ISR_TXE_TXFNF_Msk                  /*!< Transmit Data Register Empty/TXFIFO Not Full */
+#define USART_ISR_TXE                 USART_ISR_TXE_TXFNF                      /*!< BUGFIX Equivalent declaration for L4 and L4+ serial driver compatibility*/
 #define USART_ISR_LBDF_Pos            (8U)
 #define USART_ISR_LBDF_Msk            (0x1UL << USART_ISR_LBDF_Pos)            /*!< 0x00000100 */
 #define USART_ISR_LBDF                USART_ISR_LBDF_Msk                       /*!< LIN Break Detection Flag */
