diff -ruN gdb-16.3-old/gdbsupport/common-defs.h gdb-16.3/gdbsupport/common-defs.h
--- gdb-16.3-old/gdbsupport/common-defs.h	2025-04-20 19:22:06.000000000 +0200
+++ gdb-16.3/gdbsupport/common-defs.h	2026-03-11 13:46:49.640875571 +0100
@@ -82,12 +82,12 @@
    NOTE: this must be kept in sync with common.m4.  */
 #if defined (__MINGW32__) || defined (__CYGWIN__)
 # ifdef _WIN32_WINNT
-#  if _WIN32_WINNT < 0x0501
+#  if _WIN32_WINNT < 0x0600
 #   undef _WIN32_WINNT
-#   define _WIN32_WINNT 0x0501
+#   define _WIN32_WINNT 0x0600
 #  endif
 # else
-#  define _WIN32_WINNT 0x0501
+#  define _WIN32_WINNT 0x0600
 # endif
 #endif	/* __MINGW32__ || __CYGWIN__ */
 
diff -ruN gdb-16.3-old/gdbsupport/common.m4 gdb-16.3/gdbsupport/common.m4
--- gdb-16.3-old/gdbsupport/common.m4	2025-04-20 19:22:06.000000000 +0200
+++ gdb-16.3/gdbsupport/common.m4	2026-03-11 13:46:52.378870916 +0100
@@ -133,12 +133,12 @@
   dnl NOTE: this must be kept in sync with common-defs.h.
   [[#if defined (__MINGW32__) || defined (__CYGWIN__)
     # ifdef _WIN32_WINNT
-    #  if _WIN32_WINNT < 0x0501
+    #  if _WIN32_WINNT < 0x0600
     #   undef _WIN32_WINNT
-    #   define _WIN32_WINNT 0x0501
+    #   define _WIN32_WINNT 0x0600
     #  endif
     # else
-    #  define _WIN32_WINNT 0x0501
+    #  define _WIN32_WINNT 0x0600
     # endif
     #endif	/* __MINGW32__ || __CYGWIN__ */
     #include <thread>
diff -ruN gdb-16.3-old/zlib/zutil.h gdb-16.3/zlib/zutil.h
--- gdb-16.3-old/zlib/zutil.h	2024-12-29 03:50:08.000000000 +0100
+++ gdb-16.3/zlib/zutil.h	2026-03-11 09:43:49.519712177 +0100
@@ -139,15 +139,6 @@
 
 #if defined(MACOS) || defined(TARGET_OS_MAC)
 #  define OS_CODE  7
-#  ifndef Z_SOLO
-#    if defined(__MWERKS__) && __dest_os != __be_os && __dest_os != __win32_os
-#      include <unix.h> /* for fdopen */
-#    else
-#      ifndef fdopen
-#        define fdopen(fd,mode) NULL /* No fdopen() */
-#      endif
-#    endif
-#  endif
 #endif
 
 #ifdef __acorn
