diff -Naur Include.orig/stm32l100xb.h Include/stm32l100xb.h
--- Include.orig/stm32l100xb.h	2023-04-28 15:49:04
+++ Include/stm32l100xb.h	2024-01-09 15:03:15
@@ -130,7 +130,7 @@
   * @}
   */
 
-#include "core_cm3.h"
+#include <CMSIS/Include/core_cm3.h>
 #include "system_stm32l1xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l100xba.h Include/stm32l100xba.h
--- Include.orig/stm32l100xba.h	2023-04-28 15:49:04
+++ Include/stm32l100xba.h	2024-01-09 15:03:15
@@ -130,7 +130,7 @@
   * @}
   */
 
-#include "core_cm3.h"
+#include <CMSIS/Include/core_cm3.h>
 #include "system_stm32l1xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l100xc.h Include/stm32l100xc.h
--- Include.orig/stm32l100xc.h	2023-04-28 15:49:04
+++ Include/stm32l100xc.h	2024-01-09 15:03:15
@@ -137,7 +137,7 @@
   * @}
   */
 
-#include "core_cm3.h"
+#include <CMSIS/Include/core_cm3.h>
 #include "system_stm32l1xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l151xb.h Include/stm32l151xb.h
--- Include.orig/stm32l151xb.h	2023-04-28 15:49:04
+++ Include/stm32l151xb.h	2024-01-09 15:03:15
@@ -129,7 +129,7 @@
   * @}
   */
 
-#include "core_cm3.h"
+#include <CMSIS/Include/core_cm3.h>
 #include "system_stm32l1xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l151xba.h Include/stm32l151xba.h
--- Include.orig/stm32l151xba.h	2023-04-28 15:49:04
+++ Include/stm32l151xba.h	2024-01-09 15:03:15
@@ -129,7 +129,7 @@
   * @}
   */
 
-#include "core_cm3.h"
+#include <CMSIS/Include/core_cm3.h>
 #include "system_stm32l1xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l151xc.h Include/stm32l151xc.h
--- Include.orig/stm32l151xc.h	2023-04-28 15:49:04
+++ Include/stm32l151xc.h	2024-01-09 15:03:15
@@ -137,7 +137,7 @@
   * @}
   */
 
-#include "core_cm3.h"
+#include <CMSIS/Include/core_cm3.h>
 #include "system_stm32l1xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l151xca.h Include/stm32l151xca.h
--- Include.orig/stm32l151xca.h	2023-04-28 15:49:04
+++ Include/stm32l151xca.h	2024-01-09 15:03:15
@@ -137,7 +137,7 @@
   * @}
   */
 
-#include "core_cm3.h"
+#include <CMSIS/Include/core_cm3.h>
 #include "system_stm32l1xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l151xd.h Include/stm32l151xd.h
--- Include.orig/stm32l151xd.h	2023-04-28 15:49:04
+++ Include/stm32l151xd.h	2024-01-09 15:03:15
@@ -140,7 +140,7 @@
   * @}
   */
 
-#include "core_cm3.h"
+#include <CMSIS/Include/core_cm3.h>
 #include "system_stm32l1xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l151xdx.h Include/stm32l151xdx.h
--- Include.orig/stm32l151xdx.h	2023-04-28 15:49:04
+++ Include/stm32l151xdx.h	2024-01-09 15:03:15
@@ -139,7 +139,7 @@
   * @}
   */
 
-#include "core_cm3.h"
+#include <CMSIS/Include/core_cm3.h>
 #include "system_stm32l1xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l151xe.h Include/stm32l151xe.h
--- Include.orig/stm32l151xe.h	2023-04-28 15:49:04
+++ Include/stm32l151xe.h	2024-01-09 15:03:15
@@ -139,7 +139,7 @@
   * @}
   */
 
-#include "core_cm3.h"
+#include <CMSIS/Include/core_cm3.h>
 #include "system_stm32l1xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l152xb.h Include/stm32l152xb.h
--- Include.orig/stm32l152xb.h	2023-04-28 15:49:04
+++ Include/stm32l152xb.h	2024-01-09 15:03:15
@@ -130,7 +130,7 @@
   * @}
   */
 
-#include "core_cm3.h"
+#include <CMSIS/Include/core_cm3.h>
 #include "system_stm32l1xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l152xba.h Include/stm32l152xba.h
--- Include.orig/stm32l152xba.h	2023-04-28 15:49:04
+++ Include/stm32l152xba.h	2024-01-09 15:03:15
@@ -130,7 +130,7 @@
   * @}
   */
 
-#include "core_cm3.h"
+#include <CMSIS/Include/core_cm3.h>
 #include "system_stm32l1xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l152xc.h Include/stm32l152xc.h
--- Include.orig/stm32l152xc.h	2023-04-28 15:49:04
+++ Include/stm32l152xc.h	2024-01-09 15:03:15
@@ -138,7 +138,7 @@
   * @}
   */
 
-#include "core_cm3.h"
+#include <CMSIS/Include/core_cm3.h>
 #include "system_stm32l1xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l152xca.h Include/stm32l152xca.h
--- Include.orig/stm32l152xca.h	2023-04-28 15:49:04
+++ Include/stm32l152xca.h	2024-01-09 15:03:15
@@ -138,7 +138,7 @@
   * @}
   */
 
-#include "core_cm3.h"
+#include <CMSIS/Include/core_cm3.h>
 #include "system_stm32l1xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l152xd.h Include/stm32l152xd.h
--- Include.orig/stm32l152xd.h	2023-04-28 15:49:04
+++ Include/stm32l152xd.h	2024-01-09 15:03:15
@@ -141,7 +141,7 @@
   * @}
   */
 
-#include "core_cm3.h"
+#include <CMSIS/Include/core_cm3.h>
 #include "system_stm32l1xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l152xdx.h Include/stm32l152xdx.h
--- Include.orig/stm32l152xdx.h	2023-04-28 15:49:04
+++ Include/stm32l152xdx.h	2024-01-09 15:03:15
@@ -140,7 +140,7 @@
   * @}
   */
 
-#include "core_cm3.h"
+#include <CMSIS/Include/core_cm3.h>
 #include "system_stm32l1xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l152xe.h Include/stm32l152xe.h
--- Include.orig/stm32l152xe.h	2023-04-28 15:49:04
+++ Include/stm32l152xe.h	2024-01-09 15:03:15
@@ -140,7 +140,7 @@
   * @}
   */
 
-#include "core_cm3.h"
+#include <CMSIS/Include/core_cm3.h>
 #include "system_stm32l1xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l162xc.h Include/stm32l162xc.h
--- Include.orig/stm32l162xc.h	2023-04-28 15:49:04
+++ Include/stm32l162xc.h	2024-01-09 15:03:15
@@ -139,7 +139,7 @@
   * @}
   */
 
-#include "core_cm3.h"
+#include <CMSIS/Include/core_cm3.h>
 #include "system_stm32l1xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l162xca.h Include/stm32l162xca.h
--- Include.orig/stm32l162xca.h	2023-04-28 15:49:04
+++ Include/stm32l162xca.h	2024-01-09 15:03:15
@@ -139,7 +139,7 @@
   * @}
   */
 
-#include "core_cm3.h"
+#include <CMSIS/Include/core_cm3.h>
 #include "system_stm32l1xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l162xd.h Include/stm32l162xd.h
--- Include.orig/stm32l162xd.h	2023-04-28 15:49:04
+++ Include/stm32l162xd.h	2024-01-09 15:03:15
@@ -142,7 +142,7 @@
   * @}
   */
 
-#include "core_cm3.h"
+#include <CMSIS/Include/core_cm3.h>
 #include "system_stm32l1xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l162xdx.h Include/stm32l162xdx.h
--- Include.orig/stm32l162xdx.h	2023-04-28 15:49:04
+++ Include/stm32l162xdx.h	2024-01-09 15:03:15
@@ -141,7 +141,7 @@
   * @}
   */
 
-#include "core_cm3.h"
+#include <CMSIS/Include/core_cm3.h>
 #include "system_stm32l1xx.h"
 #include <stdint.h>
 
diff -Naur Include.orig/stm32l162xe.h Include/stm32l162xe.h
--- Include.orig/stm32l162xe.h	2023-04-28 15:49:04
+++ Include/stm32l162xe.h	2024-01-09 15:03:15
@@ -141,7 +141,7 @@
   * @}
   */
 
-#include "core_cm3.h"
+#include <CMSIS/Include/core_cm3.h>
 #include "system_stm32l1xx.h"
 #include <stdint.h>
 
